{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "###Differential expression analysis of PFC\n",
    "## Andi Liu\n",
    "# 1/9/2024\n",
    "\n",
    "## loading\n",
    "suppressPackageStartupMessages({\n",
    "library(Seurat)\n",
    "library(Signac)\n",
    "library(dplyr)\n",
    "library(patchwork)\n",
    "library(future)\n",
    "library(stringr)\n",
    "library(tidydr)\n",
    "library(tidyverse)\n",
    "library(viridis)\n",
    "library(qs)\n",
    "library(ggplot2)\n",
    "library(readxl)\n",
    "\n",
    "library(EnhancedVolcano)\n",
    "library(RColorBrewer)\n",
    "library(ComplexHeatmap)\n",
    "library(circlize)\n",
    "\n",
    "library(GenomicRanges)\n",
    "library(GenomicFeatures)\n",
    "library(EnsDb.Hsapiens.v86)\n",
    "library(BSgenome.Hsapiens.UCSC.hg38)\n",
    "})\n",
    "setwd(\"/data2/aliu8/2023_AD_multiome/Analysis/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### loading object\n",
    "clean <- readRDS(\"03.clean_object_rmOutlier.rds\")\n",
    "ls()\n",
    "#clean\n",
    "meta <- clean@meta.data\n",
    "head(meta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## identified replicated DEGs\n",
    "# mast\n",
    "mast_reseult <- read.csv(\"./Results/DEG/MAST_major.cell.type_PFC.csv\")\n",
    "mast_sig <- mast_reseult[mast_reseult$p_val_adj <= 0.05 & abs(mast_reseult$avg_log2FC) >= 0.25,]\n",
    "mast_sig$comb <- paste(mast_sig$gene,mast_sig$dir,mast_sig$cell.type, sep = \"_\")\n",
    "table(mast_sig$cell.type,mast_sig$dir)\n",
    "\n",
    "# mixed model\n",
    "ast <- read.csv(\"./Results/DEG/DE_mixed_PFC_Ast.csv\")\n",
    "ex <- read.csv(\"./Results/DEG/DE_mixed_PFC_Ex.csv\")\n",
    "ind <- read.csv(\"./Results/DEG/DE_mixed_PFC_In.csv\")\n",
    "mic <- read.csv(\"./Results/DEG/DE_mixed_PFC_Mic.csv\")\n",
    "opc <- read.csv(\"./Results/DEG/DE_mixed_PFC_OPC.csv\")\n",
    "oli <- read.csv(\"./Results/DEG/DE_mixed_PFC_Oli.csv\")\n",
    "# combine\n",
    "mixed_model_reseult <- rbind(ast,ex,ind,mic,opc,oli)\n",
    "mixed_model_reseult$comb <- paste(mixed_model_reseult$genes,mixed_model_reseult$dir,mixed_model_reseult$cell.type, sep = \"_\")\n",
    "write.csv(mixed_model_reseult, file = \"./Results/DEG/Mixed_PFC_major.cell.type.csv\")\n",
    "\n",
    "# filter\n",
    "mixed_sig <- mixed_model_reseult[mixed_model_reseult$p_val_adj <= 0.05 & abs(mixed_model_reseult$avg_log2FC) >=0.25,]\n",
    "table(mixed_sig$cell.type, mixed_sig$dir)\n",
    "length(mixed_model_reseult[mixed_model_reseult$p_val_adj <= 0.01 & abs(mixed_model_reseult$avg_log2FC) >=0.25,]$genes)\n",
    "length(mixed_model_reseult[mixed_model_reseult$p_val_adj <= 0.01 & abs(mixed_model_reseult$avg_log2FC) >=0.5,]$genes)\n",
    "\n",
    "# finding the overlapped signals\n",
    "sig <- mast_sig[mast_sig$comb %in% mixed_sig$comb,]\n",
    "table(sig$cell.type, sig$dir)\n",
    "length(sig$gene)\n",
    "write.csv(sig,\"./Results/DEG/Overlap_mast_mixed_EC.csv\",row.names = F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## identified replicated DEGs\n",
    "# EC\n",
    "# mast\n",
    "mast_reseult <- read.csv(\"./Results/DEG/MAST_major.cell.type_EC.csv\")\n",
    "mast_sig <- mast_reseult[mast_reseult$p_val_adj <= 0.05 & abs(mast_reseult$avg_log2FC) >= 0.25,]\n",
    "mast_sig$comb <- paste(mast_sig$gene,mast_sig$dir,mast_sig$cell.type, sep = \"_\")\n",
    "table(mast_sig$cell.type,mast_sig$dir)\n",
    "\n",
    "# mixed model\n",
    "ast <- read.csv(\"./Results/DEG/DE_mixed_EC_Ast.csv\")\n",
    "ex <- read.csv(\"./Results/DEG/DE_mixed_EC_Ex.csv\")\n",
    "ind <- read.csv(\"./Results/DEG/DE_mixed_EC_In.csv\")\n",
    "mic <- read.csv(\"./Results/DEG/DE_mixed_EC_Mic.csv\")\n",
    "opc <- read.csv(\"./Results/DEG/DE_mixed_EC_OPC.csv\")\n",
    "oli <- read.csv(\"./Results/DEG/DE_mixed_EC_Oli.csv\")\n",
    "# combine\n",
    "mixed_model_reseult <- rbind(ast,ex,ind,mic,opc,oli)\n",
    "mixed_model_reseult$comb <- paste(mixed_model_reseult$genes,mixed_model_reseult$dir,mixed_model_reseult$cell.type, sep = \"_\")\n",
    "write.csv(mixed_model_reseult, file = \"./Results/DEG/Mixed_EC_major.cell.type.csv\")\n",
    "# filter\n",
    "mixed_sig <- mixed_model_reseult[mixed_model_reseult$p_val_adj <= 0.05 & abs(mixed_model_reseult$avg_log2FC) >=0.25,]\n",
    "table(mixed_sig$cell.type, mixed_sig$dir)\n",
    "length(mixed_model_reseult[mixed_model_reseult$p_val_adj <= 0.01 & abs(mixed_model_reseult$avg_log2FC) >=0.25,]$genes)\n",
    "length(mixed_model_reseult[mixed_model_reseult$p_val_adj <= 0.01 & abs(mixed_model_reseult$avg_log2FC) >=0.5,]$genes)\n",
    "\n",
    "# finding the overlapped signals\n",
    "sig <- mast_sig[mast_sig$comb %in% mixed_sig$comb,]\n",
    "table(sig$cell.type, sig$dir)\n",
    "length(sig$gene)\n",
    "write.csv(sig,\"./Results/DEG/Overlap_mast_mixed_EC.csv\",row.names = F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## identified replicated DEGs\n",
    "# HIP\n",
    "# mast\n",
    "mast_reseult <- read.csv(\"./Results/DEG/MAST_major.cell.type_HIP.csv\")\n",
    "mast_sig <- mast_reseult[mast_reseult$p_val_adj <= 0.05 & abs(mast_reseult$avg_log2FC) >= 0.25,]\n",
    "mast_sig$comb <- paste(mast_sig$gene,mast_sig$dir,mast_sig$cell.type, sep = \"_\")\n",
    "table(mast_sig$cell.type,mast_sig$dir)\n",
    "\n",
    "# mixed model\n",
    "ast <- read.csv(\"./Results/DEG/DE_mixed_HIP_Ast.csv\")\n",
    "ex <- read.csv(\"./Results/DEG/DE_mixed_HIP_Ex.csv\")\n",
    "ind <- read.csv(\"./Results/DEG/DE_mixed_HIP_In.csv\")\n",
    "mic <- read.csv(\"./Results/DEG/DE_mixed_HIP_Mic.csv\")\n",
    "opc <- read.csv(\"./Results/DEG/DE_mixed_HIP_OPC.csv\")\n",
    "oli <- read.csv(\"./Results/DEG/DE_mixed_HIP_Oli.csv\")\n",
    "# combine\n",
    "mixed_model_reseult <- rbind(ast,ex,ind,mic,opc,oli)\n",
    "mixed_model_reseult$comb <- paste(mixed_model_reseult$genes,mixed_model_reseult$dir,mixed_model_reseult$cell.type, sep = \"_\")\n",
    "write.csv(mixed_model_reseult, file = \"./Results/DEG/Mixed_HIP_major.cell.type.csv\")\n",
    "# filter\n",
    "mixed_sig <- mixed_model_reseult[mixed_model_reseult$p_val_adj <= 0.05 & abs(mixed_model_reseult$avg_log2FC) >=0.25,]\n",
    "table(mixed_sig$cell.type, mixed_sig$dir)\n",
    "length(mixed_model_reseult[mixed_model_reseult$p_val_adj <= 0.05 & abs(mixed_model_reseult$avg_log2FC) >=0.25,]$genes)\n",
    "length(mixed_model_reseult[mixed_model_reseult$p_val_adj <= 0.05 & abs(mixed_model_reseult$avg_log2FC) >=0.5,]$genes)\n",
    "\n",
    "# finding the overlapped signals\n",
    "sig <- mast_sig[mast_sig$comb %in% mixed_sig$comb,]\n",
    "table(sig$cell.type, sig$dir)\n",
    "length(sig$gene)\n",
    "write.csv(sig,\"./Results/DEG/Overlap_mast_mixed_HIP.csv\",row.names = F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "### check overlap between three regions\n",
    "deg_pfc <- read.csv(\"./Results/DEG/Overlap_mast_mixed_PFC.csv\")\n",
    "deg_ec <- read.csv(\"./Results/DEG/Overlap_mast_mixed_EC.csv\")\n",
    "deg_hip <- read.csv(\"./Results/DEG/Overlap_mast_mixed_HIP.csv\")\n",
    "\n",
    "length(deg_pfc$gene)\n",
    "length(deg_ec$gene)\n",
    "length(deg_hip$gene)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "### Figure 2 panel a: Number of sEOAD DEGs\n",
    "df1 <- as.data.frame(table(deg_pfc$celltype,deg_pfc$dir))\n",
    "df1$region <- \"PFC\"\n",
    "\n",
    "df2 <- as.data.frame(table(deg_ec$celltype,deg_ec$dir))\n",
    "    df2$region <- \"EC\"\n",
    "\n",
    "df3 <- as.data.frame(table(deg_hip$celltype,deg_hip$dir))\n",
    "df3$region <- \"HIP\"\n",
    "#table(deg_ec$cell.type,deg_ec$dir)\n",
    "#table(deg_hip$cell.type,deg_hip$dir)\n",
    "df <- rbind(df1,df2,df3)\n",
    "colnames(df) <- c(\"celltype\",\"dir\",\"freq\",\"region\")\n",
    "\n",
    "df_pos <- df[df$dir == \"pos\",]\n",
    "df_pos<- within(df_pos, region <- factor(df_pos$region, levels=c(\"PFC\",\"EC\",\"HIP\")))\n",
    "df_neg <- df[df$dir == \"neg\",]\n",
    "df_neg<- within(df_neg, region <- factor(df_neg$region, levels=c(\"PFC\",\"EC\",\"HIP\")))\n",
    "\n",
    "# Define a color palette for regions\n",
    "region_colors <- c(\"PFC\" = \"#c25757ff\", \"EC\" = \"#825ca6ff\", \"HIP\" = \"#3f78c199\")\n",
    "region_colors\n",
    "\n",
    "generate_shades <- function(base_color, num_shades) {\n",
    "  alphas <- seq(0.2, 1, length.out = num_shades)\n",
    "  shades <- rep(base_color, num_shades)\n",
    "  shades <- scales::alpha(shades, alpha = alphas)\n",
    "  return(shades)\n",
    "}\n",
    "\n",
    "# Create a vector of colors for upregulated genes\n",
    "upregulated_colors <- generate_shades(region_colors[\"HIP\"], num_shades = 2)\n",
    "upregulated_colors\n",
    "\n",
    "# Create a vector of colors for downregulated genes (use a different shade)\n",
    "downregulated_colors <- generate_shades(region_colors[\"HIP\"], num_shades = 2)\n",
    "downregulated_colors\n",
    "\n",
    "# Grouped\n",
    "p_pos <- ggplot(df_pos, aes(fill=region, y=freq, x=celltype)) + \n",
    "         geom_bar(position=\"dodge\", stat=\"identity\") + \n",
    "         scale_fill_manual(values=c(\"#C25757\", \n",
    "                                    \"#825ca6\", \n",
    "                                    \"#3F78C1\")) +\n",
    "         ylab(\"Number of upregulated genes in EOAD\") +\n",
    "         geom_text(aes(label=freq), position=position_dodge(width = 0.95), vjust=-.5,size =6)+\n",
    "         theme(line = element_line(),\n",
    "               axis.title.y = element_text(size = 15),\n",
    "               axis.ticks.x = element_blank(),\n",
    "               axis.title.x = element_blank(),\n",
    "               axis.line.y = element_line(linewidth = 1),\n",
    "               axis.text.y = element_text(size = 15),\n",
    "               plot.background = element_blank(),\n",
    "               panel.background = element_blank(),\n",
    "               panel.grid.major = element_blank(),\n",
    "               panel.grid.minor = element_blank()\n",
    "              )\n",
    "\n",
    "p_neg <- ggplot(df_neg, aes(fill=region, y=freq, x=celltype)) + \n",
    "         geom_bar(position=\"dodge\", stat=\"identity\") + \n",
    "         scale_fill_manual(values=c(\"#C2575788\", \n",
    "                                    \"#825ca688\", \n",
    "                                    \"#3F78C188\")) +\n",
    "         ylab(\"Number of downregulated genes in EOAD\") +\n",
    "         scale_y_reverse()+\n",
    "         geom_text(aes(label=freq), position=position_dodge(width = 0.95), vjust=1.3,size =6)+\n",
    "         theme(line = element_line(),axis.text.x.bottom = element_blank(),\n",
    "               axis.title.y = element_text(size = 15),\n",
    "               axis.ticks.x = element_blank(),\n",
    "               axis.title.x = element_blank(),\n",
    "               axis.line.y = element_line(linewidth = 1),\n",
    "               axis.text.y = element_text(size = 15),\n",
    "               plot.background = element_blank(),\n",
    "               panel.background = element_blank(),\n",
    "               panel.grid.major = element_blank(),\n",
    "               panel.grid.minor = element_blank()\n",
    "              )\n",
    "pdf(\"./Figures/DEG/Number_of_DEGs.pdf\",width = 16,height =10)\n",
    "patchwork::wrap_plots(p_pos,p_neg, ncol = 1)\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Figure 2 panel b\n",
    "# jitter plot codes\n",
    "library(ggplot2)\n",
    "library(ggrepel)\n",
    "library(dplyr)\n",
    "\n",
    "jitter_deg <- function(deg_mast_path,deg_mixed_path){\n",
    "    set.seed(42)\n",
    "    ## merging two results for plotting\n",
    "    deg_mast <- read.csv(deg_mast_path,row.names = 1)\n",
    "    deg_mixed <- read.csv(deg_mixed_path)\n",
    "\n",
    "    deg_mast$sig <- ifelse(abs(deg_mast$avg_log2FC) > 0.25 & deg_mast$p_val_adj <0.05, T, F)\n",
    "    deg_mixed$sig <- ifelse(abs(deg_mixed$avg_log2FC) > 0.25 & deg_mixed$p_val_adj <0.05, T, F)\n",
    "\n",
    "    deg_mast$comb <- paste(deg_mast$gene,deg_mast$cell.type, sep = \"_\")\n",
    "    deg_mixed$comb <- paste(deg_mixed$genes,deg_mixed$cell.type, sep = \"_\")\n",
    "\n",
    "    deg <- merge(deg_mast,deg_mixed, by = \"comb\")\n",
    "    colnames(deg) <- gsub(\"\\\\.x\",\"_mast\",colnames(deg))\n",
    "    colnames(deg) <-gsub(\"\\\\.y\",\"_mixed\",colnames(deg))\n",
    "\n",
    "    deg$sig <- ifelse(deg$sig_mast == T & deg$sig_mixed == T, T, F)\n",
    "\n",
    "    ##head(deg)\n",
    "    ### top up & down for labeling\n",
    "    deg %>%\n",
    "        group_by(cell.type_mast) %>%\n",
    "        filter(sig == T) %>%\n",
    "        top_n(n = 5, wt = avg_log2FC_mast) %>%\n",
    "        filter(avg_log2FC_mast > 1)-> top_up\n",
    "    \n",
    "    deg %>%\n",
    "        group_by(cell.type_mast) %>%\n",
    "        filter(sig == T) %>%\n",
    "        top_n(n = 5, wt = -avg_log2FC_mast) %>%\n",
    "        filter(avg_log2FC_mast < -1)-> top_down\n",
    "    \n",
    "    deg_sig <- rbind(top_up, top_down)\n",
    "    deg$label <- ifelse(deg$comb %in% deg_sig$comb, T, F)\n",
    "    \n",
    "    deg_sig <- deg[deg$label == T,]\n",
    "\n",
    "   \n",
    "    pos <- position_jitter(width = 0.4, seed = 1)\n",
    "    thred <- 1\n",
    "\n",
    "    p <- ggplot(deg, aes(x=cell.type_mast, y=avg_log2FC_mast)) + \n",
    "        geom_jitter(data=subset(deg,sig == F),color='grey', size=0.5,position = pos) +\n",
    "        geom_jitter(data=subset(deg,sig == T & abs(avg_log2FC_mast)<thred),color='orange', size=0.8,position = pos) +\n",
    "        geom_jitter(data=subset(deg,sig == T & abs(avg_log2FC_mast)>thred),color='red', size=0.8,position = pos) +\n",
    "        geom_jitter(data=subset(deg,label == T & abs(avg_log2FC_mast)>thred),color='red', size=0.8, position=pos) +\n",
    "        geom_hline(yintercept=thred, linetype=\"dashed\", color = \"red\") +\n",
    "        geom_hline(yintercept=-thred, linetype=\"dashed\", color = \"red\") +\n",
    "        geom_text_repel(data=deg_sig, label=deg_sig$gene, position=pos, size=3, max.overlaps=15,min.segment.length = 0,fontface=\"italic\") +\n",
    "        theme(panel.border = element_rect(fill = \"#00000000\",linewidth = 0.5),\n",
    "            axis.title = element_blank(),\n",
    "            panel.background = element_blank(),\n",
    "            panel.grid.minor.y = element_line(color = \"#6e6e6e29\", linewidth = 0.2),\n",
    "            panel.grid.major.x = element_line(color = \"#6e6e6e29\", linewidth = 0.2),\n",
    "            axis.ticks.x = element_blank())\n",
    "    return(p)\n",
    "}\n",
    "\n",
    "## plotting\n",
    "deg_mast_pfc <- \"./Results/DEG/MAST_major.cell.type_PFC.csv\"\n",
    "deg_mixed_pfc <- \"./Results/DEG/Mixed_PFC_major.cell.type.csv\"\n",
    "p1 <- jitter_deg(deg_mast_pfc,deg_mixed_pfc)\n",
    "\n",
    "deg_mast_ec <- \"./Results/DEG/MAST_major.cell.type_EC.csv\"\n",
    "deg_mixed_ec <- \"./Results/DEG/Mixed_EC_major.cell.type.csv\"\n",
    "p2 <- jitter_deg(deg_mast_ec,deg_mixed_ec)\n",
    "\n",
    "deg_mast_hip <- \"./Results/DEG/MAST_major.cell.type_HIP.csv\"\n",
    "deg_mixed_hip <- \"./Results/DEG/Mixed_HIP_major.cell.type.csv\"\n",
    "p3 <- jitter_deg(deg_mast_hip,deg_mixed_hip)\n",
    "\n",
    "pdf(file = \"./Figures/DEG/Jitterplot_DEG_1.pdf\",width = 12,height = 8/3)\n",
    "patchwork::wrap_plots(p1,ncol = 1)\n",
    "dev.off()\n",
    "\n",
    "pdf(file = \"./Figures/DEG/Jitterplot_DEG_2.pdf\",width = 12,height = 8/3)\n",
    "patchwork::wrap_plots(p2,ncol = 1)\n",
    "dev.off()\n",
    "\n",
    "pdf(file = \"./Figures/DEG/Jitterplot_DEG_3.pdf\",width = 12,height = 8/3)\n",
    "patchwork::wrap_plots(p3,ncol = 1)\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "### Figure 2 panel c: Venn diagram across three regions\n",
    "list1 <- deg_pfc$comb\n",
    "list2 <- deg_ec$comb\n",
    "list3 <- deg_hip$comb\n",
    "\n",
    "library(VennDiagram)\n",
    "library(RColorBrewer)\n",
    "myCol <- c(\"#C25757\", \"#825ca6\", \"#3F78C1\")#brewer.pal(3, \"Pastel2\")\n",
    "\n",
    "venn.diagram(\n",
    "  x = list(list1, list2, list3),\n",
    "  category.names = c(\"sEOAD-DEGs in PFC\" , \"sEOAD-DEGs in EC\" , \"sEOAD-DEGs in HIP\"),\n",
    "  filename = './Figures/DEG/Venn_DEG_three_regions.png',\n",
    "  output=F,\n",
    "  # Output features\n",
    "  imagetype=\"png\" ,\n",
    "  height = 800, \n",
    "  width = 800, \n",
    "  resolution = 300,\n",
    "  compression = \"lzw\",\n",
    "  \n",
    "  # Circles\n",
    "  lwd = 2,\n",
    "  lty = 'blank',\n",
    "  fill = myCol,\n",
    "  \n",
    "  # Numbers\n",
    "  cex = .6,\n",
    "  fontface = \"bold\",\n",
    "  fontfamily = \"sans\",\n",
    "  \n",
    "  # Set names\n",
    "  cat.cex = 0,\n",
    "  cat.fontface = \"bold\",\n",
    "  cat.default.pos = \"outer\",\n",
    "  cat.pos = c(-27, 27, 135),\n",
    "  cat.dist = c(0.055, 0.055, 0.085),\n",
    "  cat.fontfamily = \"sans\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Figure 2 panel e/f: barplot of selected DEGs\n",
    "\n",
    "# working on EC/HIP/PFC\n",
    "# retrive metadata\n",
    "\n",
    "# CACNA1A in microglia\n",
    "DefaultAssay(object)\n",
    "meta <- object@meta.data\n",
    "table(meta$cluster_celltype)\n",
    "meta <- meta[meta$cluster_celltype == \"Cicroglia\",]\n",
    "meta$rg_diagnosis <- paste(meta$regions, meta$diagnosis, sep = \"_\")\n",
    "\n",
    "# creating dataframe for plotting\n",
    "temp <- as.data.frame(LayerData(object,layer = \"data\",assay = \"PC\",features = \"CACNA1A\",cells = rownames(meta)))\n",
    "temp <- as.data.frame(t(temp))\n",
    "\n",
    "# adding variable \n",
    "colnames(temp)[1] <- \"expression\"\n",
    "temp$diagnosis <- meta[rownames(temp),]$diagnosis\n",
    "temp$rg_diagnosis <- meta[rownames(temp),]$rg_diagnosis\n",
    "temp$rg_diagnosis <- factor(temp$rg_diagnosis,levels = c(\"PFC_EOAD\",\"PFC_NCI\",\"EC_EOAD\",\"EC_NCI\",\"HIP_EOAD\",\"HIP_NCI\"))\n",
    "\n",
    "#temp$individual_ID <- with(temp,reorder(individual_ID,diagnosis))\n",
    "p1 <- ggplot(temp, aes(x = rg_diagnosis, y = expression)) + geom_boxplot(aes(fill = diagnosis), trim=F, scale = \"width\")+theme_classic()+scale_fill_manual(values=c(\"#d40000ff\", \"#005980ff\"))+theme(legend.position = \"none\")\n",
    "\n",
    "pdf(\"./Figures/DEG/Mic_CACNA1A_boxplot.pdf\",width = 6,height = 5.3)\n",
    "p1\n",
    "dev.off()\n",
    "\n",
    "\n",
    "# GRM3 in astrocytes\n",
    "DefaultAssay(object)\n",
    "meta <- object@meta.data\n",
    "table(meta$cluster_celltype)\n",
    "meta <- meta[meta$cluster_celltype == \"Astrocyte\",]\n",
    "meta$rg_diagnosis <- paste(meta$regions, meta$diagnosis, sep = \"_\")\n",
    "\n",
    "# creating dataframe for plotting\n",
    "temp <- as.data.frame(LayerData(object,layer = \"data\",assay = \"PC\",features = \"GRM3\",cells = rownames(meta)))\n",
    "temp <- as.data.frame(t(temp))\n",
    "\n",
    "# adding variable \n",
    "colnames(temp)[1] <- \"expression\"\n",
    "temp$diagnosis <- meta[rownames(temp),]$diagnosis\n",
    "temp$rg_diagnosis <- meta[rownames(temp),]$rg_diagnosis\n",
    "temp$rg_diagnosis <- factor(temp$rg_diagnosis,levels = c(\"PFC_EOAD\",\"PFC_NCI\",\"EC_EOAD\",\"EC_NCI\",\"HIP_EOAD\",\"HIP_NCI\"))\n",
    "\n",
    "#temp$individual_ID <- with(temp,reorder(individual_ID,diagnosis))\n",
    "p2 <- ggplot(temp, aes(x = rg_diagnosis, y = expression)) + geom_boxplot(aes(fill = diagnosis), trim=F, scale = \"width\")+theme_classic()+scale_fill_manual(values=c(\"#d40000ff\", \"#005980ff\"))+theme(legend.position = \"none\")\n",
    "\n",
    "pdf(\"./Figures/DEG/Ast_GRM3_boxplot.pdf\",width = 6,height = 5.3)\n",
    "p2\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "################################################################################\n",
    "############# performing enrichment analysis on the results ####################\n",
    "################################################################################\n",
    "################################ Figure 2g #####################################\n",
    "################################################################################\n",
    "#enrichment analysis\n",
    "library(clusterProfiler)\n",
    "library('org.Hs.eg.db')\n",
    "# Package that contains MSigDB gene sets in tidy format\n",
    "library(msigdbr)\n",
    "# We will need this so we can use the pipe: %>%\n",
    "library(magrittr)\n",
    "library(DOSE)\n",
    "library(dplyr)\n",
    "library(stringr)\n",
    "library(ComplexHeatmap)\n",
    "library(circlize)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### interested in the consistent signals across regions\n",
    "### check overlap between three regions\n",
    "pfc_degs <- read.csv(\"./Results/DEG/Overlap_mast_mixed_PFC.csv\")\n",
    "ec_degs <- read.csv(\"./Results/DEG/Overlap_mast_mixed_EC.csv\")\n",
    "hip_degs <- read.csv(\"./Results/DEG/Overlap_mast_mixed_HIP.csv\")\n",
    "\n",
    "length(pfc_degs$gene)\n",
    "length(ec_degs$gene)\n",
    "length(hip_degs$gene)\n",
    "\n",
    "#head(deg_hip) \n",
    "\n",
    "## find the overlap and check\n",
    "l1 <- pfc_degs$comb\n",
    "l2 <- ec_degs$comb\n",
    "l3 <- hip_degs$comb\n",
    "\n",
    "deg_oi <- intersect(intersect(l1,l2),l3)\n",
    "sEOAD_deg <- read.csv(\"./Results/DEG/sEOAD_deg.csv\")\n",
    "deg_oi <- deg_oi[deg_oi%in%sEOAD_deg$comb]\n",
    "\n",
    "pfc_oi <- pfc_degs[pfc_degs$comb %in% deg_oi,]\n",
    "ec_oi <- ec_degs[ec_degs$comb %in% deg_oi,]\n",
    "hip_oi <- hip_degs[hip_degs$comb %in% deg_oi,]\n",
    "\n",
    "table(pfc_oi$celltype, pfc_oi$dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#########################################################################\n",
    "#### find enriched GOBP for consistent signals in inhibitory neurons ####\n",
    "#########################################################################\n",
    "pfc_degs$gene_id <- mapIds(\n",
    "  # Replace with annotation package for the organism relevant to your data\n",
    "  org.Hs.eg.db,\n",
    "  # The vector of gene identifiers we want to map\n",
    "  keys = pfc_degs$gene,\n",
    "  # Replace with the type of gene identifiers in your data\n",
    "  keytype = \"SYMBOL\",\n",
    "  # Replace with the type of gene identifiers you would like to map to\n",
    "  column = \"ENTREZID\",\n",
    "  # In the case of 1:many mappings, return the\n",
    "  # first one. This is default behavior!\n",
    "  multiVals = \"first\"\n",
    ")\n",
    "\n",
    "\n",
    "## get entrez id in data frame\n",
    "gene_key_df <- data.frame(\n",
    "  ensembl_id = pfc_degs$gene,\n",
    "  entrez_id = pfc_degs$gene_id,\n",
    "  stringsAsFactors = FALSE\n",
    ") %>%\n",
    "  # If an Ensembl gene identifier doesn't map to a gene symbol, drop that\n",
    "  # from the data frame\n",
    "  dplyr::filter(!is.na(entrez_id))\n",
    "\n",
    "degs <- pfc_degs\n",
    "head(degs)\n",
    "dim(degs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#########################################################################\n",
    "#### find enriched GOBP for consistent signals in inhibitory neurons ####\n",
    "#########################################################################\n",
    "ec_degs$gene_id <- mapIds(\n",
    "  # Replace with annotation package for the organism relevant to your data\n",
    "  org.Hs.eg.db,\n",
    "  # The vector of gene identifiers we want to map\n",
    "  keys = ec_degs$gene,\n",
    "  # Replace with the type of gene identifiers in your data\n",
    "  keytype = \"SYMBOL\",\n",
    "  # Replace with the type of gene identifiers you would like to map to\n",
    "  column = \"ENTREZID\",\n",
    "  # In the case of 1:many mappings, return the\n",
    "  # first one. This is default behavior!\n",
    "  multiVals = \"first\"\n",
    ")\n",
    "\n",
    "\n",
    "## get entrez id in data frame\n",
    "gene_key_df <- data.frame(\n",
    "  ensembl_id = ec_degs$gene,\n",
    "  entrez_id = ec_degs$gene_id,\n",
    "  stringsAsFactors = FALSE\n",
    ") %>%\n",
    "  # If an Ensembl gene identifier doesn't map to a gene symbol, drop that\n",
    "  # from the data frame\n",
    "  dplyr::filter(!is.na(entrez_id))\n",
    "\n",
    "degs <- ec_degs\n",
    "head(degs)\n",
    "dim(degs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#########################################################################\n",
    "#### find enriched GOBP for consistent signals in inhibitory neurons ####\n",
    "#########################################################################\n",
    "hip_degs$gene_id <- mapIds(\n",
    "  # Replace with annotation package for the organism relevant to your data\n",
    "  org.Hs.eg.db,\n",
    "  # The vector of gene identifiers we want to map\n",
    "  keys = hip_degs$gene,\n",
    "  # Replace with the type of gene identifiers in your data\n",
    "  keytype = \"SYMBOL\",\n",
    "  # Replace with the type of gene identifiers you would like to map to\n",
    "  column = \"ENTREZID\",\n",
    "  # In the case of 1:many mappings, return the\n",
    "  # first one. This is default behavior!\n",
    "  multiVals = \"first\"\n",
    ")\n",
    "\n",
    "\n",
    "## get entrez id in data frame\n",
    "gene_key_df <- data.frame(\n",
    "  ensembl_id = hip_degs$gene,\n",
    "  entrez_id = hip_degs$gene_id,\n",
    "  stringsAsFactors = FALSE\n",
    ") %>%\n",
    "  # If an Ensembl gene identifier doesn't map to a gene symbol, drop that\n",
    "  # from the data frame\n",
    "  dplyr::filter(!is.na(entrez_id))\n",
    "\n",
    "degs <- hip_degs\n",
    "head(degs)\n",
    "dim(degs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## create a function to ran ORA for all degs in each cell type region. Then, combine the result ##\n",
    "get_ct_enrichment <- function(celltype,deg_dir,pfc_degs,ec_degs,hip_degs){\n",
    "    ## set parameters\n",
    "    celltype = celltype\n",
    "    deg_dir = deg_dir\n",
    "    pfc_degs = pfc_degs\n",
    "    ec_degs = ec_degs\n",
    "    hip_degs = hip_degs\n",
    "    message(paste(\"Working on \",celltype,\" \", deg_dir,\".\",sep = \"\"))\n",
    "    #### enrichment analysis of degs\n",
    "    ego <- enrichGO(gene          = pfc_degs[pfc_degs$celltype==celltype & pfc_degs$dir==deg_dir,]$gene_id,\n",
    "                    OrgDb         = org.Hs.eg.db,\n",
    "                    ont           = c(\"BP\"),\n",
    "                    pAdjustMethod = \"BH\",\n",
    "                    pvalueCutoff  = 0.05,\n",
    "#                   qvalueCutoff  = 0.2,\n",
    "                    readable      = TRUE)\n",
    "    ## remove redundant results                \n",
    "    ego <- simplify(ego)\n",
    "    pfc_up<-ego@result\n",
    "\n",
    "    ego <- enrichGO(gene          = ec_degs[ec_degs$celltype == celltype & ec_degs$dir == deg_dir, ]$gene_id,\n",
    "                    OrgDb         = org.Hs.eg.db,\n",
    "                    ont           = c(\"BP\"),\n",
    "                    pAdjustMethod = \"BH\",\n",
    "                    pvalueCutoff  = 0.05,\n",
    "#                   qvalueCutoff  = 0.2,\n",
    "                    readable      = TRUE)\n",
    "    ## remove redundant results                \n",
    "    ego <- simplify(ego)\n",
    "    ec_up<-ego@result\n",
    "\n",
    "# \n",
    "    ego <- enrichGO(gene          = hip_degs[hip_degs$celltype == celltype & hip_degs$dir == deg_dir, ]$gene_id,\n",
    "                    OrgDb         = org.Hs.eg.db,\n",
    "                    ont           = c(\"BP\"),\n",
    "                    pAdjustMethod = \"BH\",\n",
    "                    pvalueCutoff  = 0.05,\n",
    "#                   qvalueCutoff  = 0.2,\n",
    "                    readable      = TRUE)\n",
    "                    \n",
    "    ## remove redundant results                \n",
    "    ego <- simplify(ego)\n",
    "    hip_up<-ego@result\n",
    "\n",
    "    ## keep all results? then also plot non-significant results\n",
    "    #pfc_up <- pfc_up[pfc_up$p.adjust <= 0.05,]\n",
    "    ifelse(nrow(pfc_up) > 0,pfc_up$region <- \"PFC\",pfc_up$region <- character())\n",
    "    #head(pfc_up)\n",
    "\n",
    "    #ec_up <- ec_up[ec_up$p.adjust <= 0.05,]\n",
    "    ifelse(nrow(ec_up) > 0,ec_up$region <- \"EC\",ec_up$region <- character())\n",
    "    #head(ec_up)\n",
    "\n",
    "    #hip_up <- hip_up[hip_up$p.adjust <= 0.05,]\n",
    "    ifelse(nrow(hip_up) > 0,hip_up$region <- \"HIP\",hip_up$region <- character())\n",
    "    #head(hip_up)\n",
    "    df_up <- data.frame()\n",
    "    df_up <- rbind(df_up,pfc_up,ec_up,hip_up)\n",
    "    df_up$percent <- as.numeric(str_split_fixed(df_up$GeneRatio,pattern = \"/\",n = 2)[,1]) / as.numeric(str_split_fixed(df_up$GeneRatio,pattern = \"/\",n = 2)[,2])\n",
    "    #terms <- intersect(intersect(pfc_ast_up$ID, ec_ast_up$ID),hip_ast_up$ID)\n",
    "\n",
    "    print(table(df_up$region))\n",
    "\n",
    "    # check how many significat\n",
    "    print(table(df_up[df_up$qvalue < 0.05,]$region))\n",
    "\n",
    "    # provide cell type information\n",
    "    df_up$celltype <- celltype\n",
    "    return(df_up)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## running enrichment analysis\n",
    "ast_up <- get_ct_enrichment(celltype = \"Astrocyte\",deg_dir = \"pos\", pfc_degs = pfc_degs,ec_degs = ec_degs,hip_degs = hip_degs)\n",
    "exc_up <- get_ct_enrichment(celltype = \"Excitatory\",deg_dir = \"pos\", pfc_degs = pfc_degs,ec_degs = ec_degs,hip_degs = hip_degs)\n",
    "inh_up <- get_ct_enrichment(celltype = \"Inhibitory\",deg_dir = \"pos\", pfc_degs = pfc_degs,ec_degs = ec_degs,hip_degs = hip_degs)\n",
    "mic_up <- get_ct_enrichment(celltype = \"Microglia\",deg_dir = \"pos\", pfc_degs = pfc_degs,ec_degs = ec_degs,hip_degs = hip_degs)\n",
    "oli_up <- get_ct_enrichment(celltype = \"Oligodendrocyte\",deg_dir = \"pos\", pfc_degs = pfc_degs,ec_degs = ec_degs,hip_degs = hip_degs)\n",
    "opc_up <- get_ct_enrichment(celltype = \"OPC\",deg_dir = \"pos\", pfc_degs = pfc_degs,ec_degs = ec_degs,hip_degs = hip_degs)\n",
    "#table(ast_up$region)\n",
    "#table(ast_up[ast_up$qvalue <0.05,]$region)\n",
    "\n",
    "## combine all df\n",
    "df <- rbind(ast_up, exc_up, inh_up, mic_up, oli_up, opc_up)\n",
    "table(df$region,df$celltype)\n",
    "#table(df[df$qvalue < 0.05,]$region,df[df$qvalue < 0.05,]$celltype)\n",
    "\n",
    "# giving unique colnames\n",
    "df$colname <- paste(substr(df$celltype,start=1,stop=3),df$region,sep = \"_\")\n",
    "\n",
    "# saving\n",
    "write.csv(df,file = \"./Results/DEG/GO_pos_deg_all_ct_regions_no_redundant.csv.csv\")\n",
    "\n",
    "df_all <- read.csv(\"./Results/DEG/GO_pos_deg_all_ct_regions.csv\",row.names = 1)\n",
    "df <- read.csv(\"./Results/DEG/GO_pos_deg_all_ct_regions_no_redundant.csv\",row.names = 1)\n",
    "\n",
    "table(df_all$region,df_all$celltype)\n",
    "table(df_all[df_all$qvalue < 0.05,]$region,df_all[df_all$qvalue < 0.05,]$celltype)\n",
    "table(df[df$qvalue < 0.05,]$region,df[df$qvalue < 0.05,]$celltype)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## plotting\n",
    "top <- df %>% group_by(colname) %>% top_n(n = 5, wt = -qvalue) %>% top_n(n = 5, wt = -pvalue)\n",
    "dim(top)\n",
    "top <- top[top$qvalue < 0.05,]\n",
    "top <- top$Description \n",
    "\n",
    "mat<-matrix(nrow=length(unique(top)), ncol=18)\n",
    "colnames(mat) <- unique(df_all$colname)\n",
    "\n",
    "for (j in 1:length(unique(top))){\n",
    "  for (i in 1:18){\n",
    "    termTemp<-unique(top)[j]\n",
    "    rgTemp<-colnames(mat)[i]\n",
    "\n",
    "    sub<-df_all[which(df_all$Description==termTemp & df_all$colname==rgTemp),]\n",
    "    if (nrow(sub)>0){\n",
    "      mat[j,i]<-sub$Count\n",
    "    }\n",
    "    else{\n",
    "      mat[j,i]<-0\n",
    "    }\n",
    "  }  \n",
    "}\n",
    "rownames(mat)<- unique(top)\n",
    "\n",
    "#mat\n",
    "\n",
    "#generating a significance matrix\n",
    "sig_mat<-matrix(nrow=length(unique(top)), ncol=18)\n",
    "colnames(sig_mat)<-unique(df_all$colname)\n",
    "\n",
    "for (j in 1:length(unique(top))){\n",
    "  for (i in 1:18){\n",
    "    termTemp<-unique(top)[j]\n",
    "    rgTemp<-colnames(sig_mat)[i]\n",
    "    sub<-df_all[which(df_all$Description==termTemp & df_all$colname==rgTemp),]\n",
    "    if (nrow(sub)>0){\n",
    "      sig_mat[j,i]<-sub$qvalue}\n",
    "    else{\n",
    "      sig_mat[j,i]<-1}  }}\n",
    "\n",
    "\n",
    "#### plotting the upregulated gene GSEA\n",
    "ha<-HeatmapAnnotation(Region=colnames(mat)\n",
    "                       , col= list(Region=c(\"Ast_PFC\"=\"#c25757ff\",\"Ast_EC\"=\"#825ca6ff\",\"Ast_HIP\"=\"#3f78c1ff\",\n",
    "                       \"Exc_PFC\"=\"#c25757ff\",\"Exc_EC\"=\"#825ca6ff\",\"Exc_HIP\"=\"#3f78c1ff\",\n",
    "                       \"Inh_PFC\"=\"#c25757ff\",\"Inh_EC\"=\"#825ca6ff\",\"Inh_HIP\"=\"#3f78c1ff\",\n",
    "                       \"Mic_PFC\"=\"#c25757ff\",\"Mic_EC\"=\"#825ca6ff\",\"Mic_HIP\"=\"#3f78c1ff\",\n",
    "                       \"Oli_PFC\"=\"#c25757ff\",\"Oli_EC\"=\"#825ca6ff\",\"Oli_HIP\"=\"#3f78c1ff\",\n",
    "                       \"OPC_PFC\"=\"#c25757ff\",\"OPC_EC\"=\"#825ca6ff\",\"OPC_HIP\"=\"#3f78c1ff\")), show_legend=F,annotation_label=\"\")\n",
    "\n",
    "ha2<-HeatmapAnnotation(celltype = colnames(mat)\n",
    "                       , col= list(celltype=c(\"Ast_PFC\"=\"#F06719\",\"Ast_EC\"=\"#F06719\",\"Ast_HIP\"=\"#F06719\",\n",
    "                       \"Exc_PFC\"=\"#33A65C\",\"Exc_EC\"=\"#33A65C\",\"Exc_HIP\"=\"#33A65C\",\n",
    "                       \"Inh_PFC\"=\"#23767C\",\"Inh_EC\"=\"#23767C\",\"Inh_HIP\"=\"#23767C\",\n",
    "                       \"Mic_PFC\"=\"#E03426\",\"Mic_EC\"=\"#E03426\",\"Mic_HIP\"=\"#E03426\",\n",
    "                       \"Oli_PFC\"=\"#1BA3C6\",\"Oli_EC\"=\"#1BA3C6\",\"Oli_HIP\"=\"#1BA3C6\",\n",
    "                       \"OPC_PFC\"=\"#A26DC2\",\"OPC_EC\"=\"#A26DC2\",\"OPC_HIP\"=\"#A26DC2\")), show_legend=F,annotation_label=\"\")\n",
    "ha <- c(ha2,ha)\n",
    "\n",
    "\n",
    "ht=Heatmap(mat, cluster_rows=F,cluster_columns=F,\n",
    "           col=colorRamp2(c(0,30,60),c(\"grey95\",\"red\",\"red4\")),  \n",
    "           top_annotation=ha, #name=\"Number of genes with significant expression changes\",  \n",
    "           show_column_names=T, show_row_names=T, column_title=NULL,\n",
    "           row_names_side=\"right\", row_title_gp=gpar(fontsize=60),\n",
    "           row_names_max_width = max_text_width(rownames(mat), gp = gpar(fontsize = 24)), \n",
    "           cell_fun = function(j, i, x, y, w, h, fill) {\n",
    "  if(sig_mat[i, j] <0.05) {\n",
    "    grid.text(\"*\", x, y, gp=gpar(fontsize=35, col=\"black\"), vjust=\"center\")\n",
    "  } })\n",
    "\n",
    "p1 <- ht\n",
    "\n",
    "pdf(\"./Figures/DEG/up_GSEA.pdf\", width=12, height=12)\n",
    "p1\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## running enrichment analysis for down regulated genes\n",
    "ast_down <- get_ct_enrichment(celltype = \"Astrocyte\",deg_dir = \"neg\", pfc_degs = pfc_degs,ec_degs = ec_degs,hip_degs = hip_degs)\n",
    "exc_down <- get_ct_enrichment(celltype = \"Excitatory\",deg_dir = \"neg\", pfc_degs = pfc_degs,ec_degs = ec_degs,hip_degs = hip_degs)\n",
    "inh_down <- get_ct_enrichment(celltype = \"Inhibitory\",deg_dir = \"neg\", pfc_degs = pfc_degs,ec_degs = ec_degs,hip_degs = hip_degs)\n",
    "mic_down <- get_ct_enrichment(celltype = \"Microglia\",deg_dir = \"neg\", pfc_degs = pfc_degs,ec_degs = ec_degs,hip_degs = hip_degs)\n",
    "oli_down <- get_ct_enrichment(celltype = \"Oligodendrocyte\",deg_dir = \"neg\", pfc_degs = pfc_degs,ec_degs = ec_degs,hip_degs = hip_degs)\n",
    "opc_down <- get_ct_enrichment(celltype = \"OPC\",deg_dir = \"neg\", pfc_degs = pfc_degs,ec_degs = ec_degs,hip_degs = hip_degs)\n",
    "#table(ast_up$region)\n",
    "#table(ast_up[ast_up$qvalue <0.05,]$region)\n",
    "\n",
    "## combine all df\n",
    "df <- rbind(ast_down, exc_down, inh_down, mic_down, oli_down, opc_down)\n",
    "table(df$region,df$celltype)\n",
    "table(df[df$qvalue < 0.05,]$region,df[df$qvalue < 0.05,]$celltype)\n",
    "\n",
    "# giving unique colnames\n",
    "df$colname <- paste(substr(df$celltype,start=1,stop=3),df$region,sep = \"_\")\n",
    "\n",
    "# saving\n",
    "write.csv(df,file = \"./Results/DEG/GO_neg_deg_all_ct_regions_no_redundant.csv\")\n",
    "\n",
    "##\n",
    "df_all <- read.csv(\"./Results/DEG/GO_neg_deg_all_ct_regions.csv\",row.names = 1)\n",
    "df <- read.csv(\"./Results/DEG/GO_neg_deg_all_ct_regions_no_redundant.csv\",row.names = 1)\n",
    "#df <- read.csv(\"./Results/DEG/GO_pos_deg_all_ct_regions.csv\",row.names = 1)\n",
    "\n",
    "table(df_all[df_all$qvalue < 0.05,]$region,df_all[df_all$qvalue < 0.05,]$celltype)\n",
    "table(df[df$qvalue < 0.05,]$region,df[df$qvalue < 0.05,]$celltype)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##n plotting \n",
    "top <- df %>% group_by(colname) %>% top_n(n = 5, wt = -qvalue) %>% top_n(n = 5, wt = -pvalue)\n",
    "dim(top)\n",
    "top <- top[top$qvalue < 0.05,]\n",
    "top <- top$Description \n",
    "\n",
    "mat<-matrix(nrow=length(unique(top)), ncol=18)\n",
    "colnames(mat) <- unique(df_all$colname)\n",
    "\n",
    "for (j in 1:length(unique(top))){\n",
    "  for (i in 1:18){\n",
    "    termTemp<-unique(top)[j]\n",
    "    rgTemp<-colnames(mat)[i]\n",
    "\n",
    "    sub<-df_all[which(df_all$Description==termTemp & df_all$colname==rgTemp),]\n",
    "    if (nrow(sub)>0){\n",
    "      mat[j,i]<-sub$Count\n",
    "    }\n",
    "    else{\n",
    "      mat[j,i]<-0\n",
    "    }\n",
    "  }  \n",
    "}\n",
    "rownames(mat)<- unique(top)\n",
    "\n",
    "#mat\n",
    "\n",
    "#generating a significance matrix\n",
    "sig_mat<-matrix(nrow=length(unique(top)), ncol=18)\n",
    "colnames(sig_mat)<-unique(df_all$colname)\n",
    "\n",
    "for (j in 1:length(unique(top))){\n",
    "  for (i in 1:18){\n",
    "    termTemp<-unique(top)[j]\n",
    "    rgTemp<-colnames(sig_mat)[i]\n",
    "    sub<-df_all[which(df_all$Description==termTemp & df_all$colname==rgTemp),]\n",
    "    if (nrow(sub)>0){\n",
    "      sig_mat[j,i]<-sub$qvalue}\n",
    "    else{\n",
    "      sig_mat[j,i]<-1}  }}\n",
    "\n",
    "#sig_mat\n",
    "\n",
    "## plotting for the results of GSEA_downregulated\n",
    "ha<-HeatmapAnnotation(Region=colnames(mat)\n",
    "                       , col= list(Region=c(\"Ast_PFC\"=\"#c25757ff\",\"Ast_EC\"=\"#825ca6ff\",\"Ast_HIP\"=\"#3f78c1ff\",\n",
    "                       \"Exc_PFC\"=\"#c25757ff\",\"Exc_EC\"=\"#825ca6ff\",\"Exc_HIP\"=\"#3f78c1ff\",\n",
    "                       \"Inh_PFC\"=\"#c25757ff\",\"Inh_EC\"=\"#825ca6ff\",\"Inh_HIP\"=\"#3f78c1ff\",\n",
    "                       \"Mic_PFC\"=\"#c25757ff\",\"Mic_EC\"=\"#825ca6ff\",\"Mic_HIP\"=\"#3f78c1ff\",\n",
    "                       \"Oli_PFC\"=\"#c25757ff\",\"Oli_EC\"=\"#825ca6ff\",\"Oli_HIP\"=\"#3f78c1ff\",\n",
    "                       \"OPC_PFC\"=\"#c25757ff\",\"OPC_EC\"=\"#825ca6ff\",\"OPC_HIP\"=\"#3f78c1ff\")), show_legend=F,annotation_label=\"\")\n",
    "\n",
    "ha2<-HeatmapAnnotation(celltype = colnames(mat)\n",
    "                       , col= list(celltype=c(\"Ast_PFC\"=\"#F06719\",\"Ast_EC\"=\"#F06719\",\"Ast_HIP\"=\"#F06719\",\n",
    "                       \"Exc_PFC\"=\"#33A65C\",\"Exc_EC\"=\"#33A65C\",\"Exc_HIP\"=\"#33A65C\",\n",
    "                       \"Inh_PFC\"=\"#23767C\",\"Inh_EC\"=\"#23767C\",\"Inh_HIP\"=\"#23767C\",\n",
    "                       \"Mic_PFC\"=\"#E03426\",\"Mic_EC\"=\"#E03426\",\"Mic_HIP\"=\"#E03426\",\n",
    "                       \"Oli_PFC\"=\"#1BA3C6\",\"Oli_EC\"=\"#1BA3C6\",\"Oli_HIP\"=\"#1BA3C6\",\n",
    "                       \"OPC_PFC\"=\"#A26DC2\",\"OPC_EC\"=\"#A26DC2\",\"OPC_HIP\"=\"#A26DC2\")), show_legend=F,annotation_label=\"\")\n",
    "ha <- c(ha2,ha)\n",
    "\n",
    "ht=Heatmap(mat, cluster_rows=F,cluster_columns=F,\n",
    "           col=colorRamp2(c(0,30,60),c(\"grey95\",\"deepskyblue\",\"dodgerblue4\")),  \n",
    "           top_annotation=ha, #name=\"Number of genes with significant expression changes\",  \n",
    "           show_column_names=T, show_row_names=T, column_title=NULL,\n",
    "           row_names_side=\"right\", row_title_gp=gpar(fontsize=60),\n",
    "           row_names_max_width = max_text_width(rownames(mat), gp = gpar(fontsize = 24)), \n",
    "           cell_fun = function(j, i, x, y, w, h, fill) {\n",
    "  if(sig_mat[i, j] <0.05) {\n",
    "    grid.text(\"*\", x, y, gp=gpar(fontsize=35, col=\"black\"), vjust=\"center\")\n",
    "  } })\n",
    "\n",
    "p1 <- ht\n",
    "\n",
    "pdf(\"./Figures/DEG/down_GSEA.pdf\", width=12, height=12)\n",
    "p1\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Supplmentary figure overlap of sEOAD DEGs in PFC with existing LOAD snRNA-seq studies\n",
    "# load PFC overlap DEG results\n",
    "sig <- read.csv(\"./Results/DEG/Overlap_mast_mixed_PFC.csv\",row.names = 1)\n",
    "length(sig$comb)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### overlap between DEG of PFC in current study and existing publications\n",
    "library(readxl)\n",
    "## Morabito et al. \n",
    "morabito <- read_excel(\"./Data/Morabito.xlsx\",sheet = 5)\n",
    "morabito <- morabito[abs(morabito$avg_logFC)>=0.25 & morabito$p_val_adj <=0.05, ]\n",
    "table(morabito$celltype)\n",
    "morabito[morabito$celltype == \"ASC\",]$celltype <- \"Astrocyte\"\n",
    "morabito[morabito$celltype == \"EX\",]$celltype <- \"Excitatory\"\n",
    "morabito[morabito$celltype == \"INH\",]$celltype <- \"Inhibitory\"\n",
    "morabito[morabito$celltype == \"MG\",]$celltype <- \"Microglia\"\n",
    "morabito[morabito$celltype == \"ODC\",]$celltype <- \"Oligodendrocyte\"\n",
    "morabito[morabito$celltype == \"OPC\",]$celltype <- \"OPC\"\n",
    "morabito <- morabito[morabito$celltype != \"PER.END\",]\n",
    "morabito$dir <- ifelse(morabito$avg_logFC >0, \"pos\",\"neg\")\n",
    "morabito$comb <- paste(morabito$gene,morabito$dir,morabito$celltype, sep = \"_\")\n",
    "\n",
    "\n",
    "## LOAD multiome Anderson et al.\n",
    "LOAD_multiome <- read_excel(\"./Data/LOAD_multiome.xlsx\",sheet = 2)\n",
    "LOAD_multiome <- LOAD_multiome[abs(LOAD_multiome$avg_log2FC)>=0.25 & LOAD_multiome$p_val_adj <=0.05, ]\n",
    "LOAD_multiome$dir <- ifelse(LOAD_multiome$cat == \"up\",\"pos\",\"neg\")\n",
    "table(LOAD_multiome$celltype)\n",
    "LOAD_multiome[LOAD_multiome$celltype == \"Astrocytes\",]$celltype <- \"Astrocyte\"\n",
    "LOAD_multiome[LOAD_multiome$celltype == \"Microglia\",]$celltype <- \"Microglia\"\n",
    "LOAD_multiome[LOAD_multiome$celltype == \"Oligodendrocytes\",]$celltype <- \"Oligodendrocyte\"\n",
    "LOAD_multiome[LOAD_multiome$celltype == \"OPCs\",]$celltype <- \"OPC\"\n",
    "LOAD_multiome$comb <- paste(LOAD_multiome$gene,LOAD_multiome$dir,LOAD_multiome$celltype, sep = \"_\")\n",
    "\n",
    "## Mathy et al. \n",
    "mathy <- read.csv(\"./Data/Mathy.csv\")\n",
    "table(mathy$celltype)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Supplmentary figure overlap of sEOAD DEGs in PFC with existing LOAD snRNA-seq studies\n",
    "## list of degs\n",
    "list1 <- sig$comb\n",
    "list2 <- morabito$comb\n",
    "list3 <- LOAD_multiome$comb\n",
    "list4 <- mathy$comb\n",
    "\n",
    "## drawing venn diagram\n",
    "library(VennDiagram)\n",
    "library(RColorBrewer)\n",
    "myCol <- brewer.pal(4, \"Pastel2\")\n",
    "\n",
    "venn.diagram(\n",
    "  x = list(l5, list2, list3,list4),\n",
    "  category.names = c(\"EOAD_PFC\" , \"Morabito et al.\" , \"Anderson et al.\",\"Mathy et al.\"),\n",
    "  filename = './Figures/Venn_DEG_PFC.png',\n",
    "  output=F,\n",
    "  # Output features\n",
    "  imagetype=\"png\" ,\n",
    "  height = 1200, \n",
    "  width = 1500, \n",
    "  resolution = 300,\n",
    "  compression = \"lzw\",\n",
    "  \n",
    "  # Circles\n",
    "  lwd = 2,\n",
    "  lty = 'blank',\n",
    "  fill = myCol,\n",
    "  \n",
    "  # Numbers\n",
    "  cex = .6,\n",
    "  fontface = \"bold\",\n",
    "  fontfamily = \"sans\",\n",
    "  \n",
    "  # Set names\n",
    "  cat.cex = 0.6,\n",
    "  cat.fontface = \"bold\",\n",
    "  cat.default.pos = \"outer\",\n",
    "#  cat.pos = c(-27, 27, 135),\n",
    "#  cat.dist = c(0.055, 0.055, 0.085),\n",
    "  cat.fontfamily = \"sans\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Supplementary figures upset plots\n",
    "### check overlap between three regions\n",
    "deg_pfc <- read.csv(\"./Results/DEG/Overlap_mast_mixed_PFC.csv\")\n",
    "deg_ec <- read.csv(\"./Results/DEG/Overlap_mast_mixed_EC.csv\")\n",
    "deg_hip <- read.csv(\"./Results/DEG/Overlap_mast_mixed_HIP.csv\")\n",
    "\n",
    "length(deg_pfc$gene)\n",
    "length(deg_ec$gene)\n",
    "length(deg_hip$gene)\n",
    "\n",
    "### function for plotting\n",
    "upset_region <- function(ct){\n",
    "    print(ct)\n",
    "    ct = ct\n",
    "    l1 <- deg_pfc[deg_pfc$celltype == ct,]$comb\n",
    "    l2 <- deg_ec[deg_ec$celltype == ct,]$comb\n",
    "    l3 <- deg_hip[deg_hip$celltype == ct,]$comb\n",
    "\n",
    "    lt = list(PFC = l1,\n",
    "              EC = l2,\n",
    "              HIP = l3)\n",
    "    print(length(lt$PFC))\n",
    "    m1 = make_comb_mat(lt)\n",
    "\n",
    "    p <- UpSet(m1,set_order = c(\"PFC\", \"EC\", \"HIP\"),comb_order = order(-comb_size(m1)),\n",
    "        top_annotation = upset_top_annotation(m1, gp = gpar(col = comb_degree(m1))),\n",
    "        row_title = ct)\n",
    "    return(p)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## plot the upset plot of each cell type.\n",
    "ct_list <- c(\"Astrocyte\",\"Excitatory\",\"Inhibitory\",\"Microglia\",\"Oligodendrocyte\",\"OPC\")\n",
    "\n",
    "p1 <- upset_region(ct_list[1])\n",
    "p2 <- upset_region(ct_list[2])\n",
    "p3 <- upset_region(ct_list[3])\n",
    "p4 <- upset_region(ct_list[4])\n",
    "p5 <- upset_region(ct_list[5])\n",
    "p6 <- upset_region(ct_list[6])\n",
    "\n",
    "## plotting\n",
    "pdf(file = \"./Figures/DEG/Upset_DEGs_Ast.pdf\",width = 4, height = 3)\n",
    "p1\n",
    "dev.off()\n",
    "\n",
    "pdf(file = \"./Figures/DEG/Upset_DEGs_Exc.pdf\",width = 4, height = 3)\n",
    "p2\n",
    "dev.off()\n",
    "\n",
    "pdf(file = \"./Figures/DEG/Upset_DEGs_Inh.pdf\",width = 4, height = 3)\n",
    "p3\n",
    "dev.off()\n",
    "\n",
    "pdf(file = \"./Figures/DEG/Upset_DEGs_Mic.pdf\",width = 4, height = 3)\n",
    "p4\n",
    "dev.off()\n",
    "\n",
    "pdf(file = \"./Figures/DEG/Upset_DEGs_Oli.pdf\",width = 4, height = 3)\n",
    "p5\n",
    "dev.off()\n",
    "\n",
    "pdf(file = \"./Figures/DEG/Upset_DEGs_Opc.pdf\",width = 4, height = 3)\n",
    "p6\n",
    "dev.off()"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
