{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# overlap linked peaks to GWAS\n",
    "library(stringr)\n",
    "library(GenomicRanges)\n",
    "library(GenomicFeatures)\n",
    "library(LDlinkR)\n",
    "\n",
    "## Loading\n",
    "library(Seurat)\n",
    "library(Signac)\n",
    "library(dplyr)\n",
    "library(patchwork)\n",
    "library(future)\n",
    "library(stringr)\n",
    "library(tidydr)\n",
    "library(tidyverse)\n",
    "library(viridis)\n",
    "library(ggplot2)\n",
    "library(EnsDb.Hsapiens.v86)\n",
    "library(BSgenome.Hsapiens.UCSC.hg38)\n",
    "\n",
    "# For motif analysis\n",
    "library(JASPAR2020)\n",
    "library(TFBSTools)\n",
    "\n",
    "##\n",
    "setwd(\"/data2/aliu8/2023_AD_multiome/Analysis/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## using linked peaks results as the input for LDSC\n",
    "list.files(\"./Results/ldsc/\")\n",
    "list.files(\"./Results/LINK/\")\n",
    "list.files(\"./Results/LINK/\")\n",
    "\n",
    "## in house CT peaks annotation files\n",
    "ctpeaks <- read.csv(\"./Results/CTpeaks_annotated.csv\",row.names = 1)\n",
    "colnames(ctpeaks) <- c(\"chr\",\"start\",\"end\",\"width\",\"strand\",\"peak_called_in\")\n",
    "ctpeaks <- GRanges(ctpeaks)\n",
    "ctpeaks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "############################################# \n",
    "### Generate bed file for S-LDSC analysis ###\n",
    "#############################################\n",
    "########### for DEG linked peaks ############\n",
    "#############################################\n",
    "pfc_linked_peaks <- read.csv(\"./Results/LINK/PFC_linkpeaks_all_annotated.csv\")\n",
    "ec_linked_peaks <- read.csv(\"./Results/LINK/EC_linkpeaks_all_annotated.csv\")\n",
    "hip_linked_peaks <- read.csv(\"./Results/LINK/HIP_linkpeaks_all_annotated.csv\")\n",
    "\n",
    "table(pfc_linked_peaks$celltype)\n",
    "table(ec_linked_peaks$celltype)\n",
    "table(hip_linked_peaks$celltype)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "##### restricting peaks called only in one cell type\n",
    "## OPTIONAL ##\n",
    "## generating cleaning cell type peaks files for six major cell types\n",
    "ct <- c(\"Astrocyte\",\"Excitatory\",\"Inhibitory\",\"Microglia\",\"Oligodendrocyte\",\"OPC\")\n",
    "# select specific peaks\n",
    "peaks <- ec_linked_peaks\n",
    "\n",
    "final <- data.frame()\n",
    "\n",
    "for(i in ct){\n",
    "    # assign cell type\n",
    "    celltype = i\n",
    "\n",
    "    ## ct peaks that are not cell type AD-DEG linked peaks\n",
    "    #anno_temp <- ctpeaks[grep(celltype,ctpeaks$peak_called_in)]\n",
    "    anno_temp <- ctpeaks[ctpeaks$peak_called_in == celltype]\n",
    "\n",
    "    # get signals within specific cell type\n",
    "    atac_peaks <- peaks[peaks$celltype == celltype,]\n",
    "    link_temp <- GRanges(atac_peaks[,c(\"seqnames\",\"peak.start\",\"peak.end\")]) #chr or seqnames\n",
    "\n",
    "    print(table(link_temp %in% anno_temp))\n",
    "    # OUTPUT filtered results\n",
    "    out <- atac_peaks[which(link_temp %in% anno_temp),]\n",
    "\n",
    "    final <- rbind(final, out)\n",
    "}\n",
    "table(final$celltype,useNA = \"always\")\n",
    "\n",
    "############################################# \n",
    "### Generate bed file for S-LDSC analysis ###\n",
    "#############################################\n",
    "############## for deg results ##############\n",
    "#############################################\n",
    "table(final$celltype)\n",
    "final <- final[order(c(final$seqnames)),]\n",
    "## generating peak region\n",
    "final$seqnames <- str_split_fixed(final$peak, \"-\",3)[,1]\n",
    "final$start <- str_split_fixed(final$peak, \"-\",3)[,2]\n",
    "final$end <- str_split_fixed(final$peak, \"-\",3)[,3]\n",
    "\n",
    "ct <- c(\"Astrocyte\",\"Excitatory\",\"Inhibitory\",\"Microglia\",\"Oligodendrocyte\",\"OPC\") #,\"OPC\"\n",
    "\n",
    "for (i in ct){\n",
    "    print(i)\n",
    "    df <- final[final$celltype == i,c(1,2,3)]\n",
    "    print(head(df))\n",
    "    write.table(\n",
    "        df, \n",
    "        file = paste(\"./Results/ldsc/DEG_linked_peaks_restricted/\",\"EC_\",i,\"_DEG_linked_peaks_restricted.bed\",sep = \"\"),\n",
    "        quote=F,row.names=F, col.names=F, sep=\"\\t\")\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "##### restricting peaks called only in one cell type\n",
    "## OPTIONAL ##\n",
    "## generating cleaning cell type peaks files for six major cell types\n",
    "ct <- c(\"Astrocyte\",\"Excitatory\",\"Inhibitory\",\"Microglia\",\"Oligodendrocyte\",\"OPC\")\n",
    "# select specific peaks\n",
    "peaks <- pfc_linked_peaks\n",
    "\n",
    "final <- data.frame()\n",
    "\n",
    "for(i in ct){\n",
    "    # assign cell type\n",
    "    celltype = i\n",
    "\n",
    "    ## ct peaks that are not cell type AD-DEG linked peaks\n",
    "    anno_temp <- ctpeaks[ctpeaks$peak_called_in == celltype]\n",
    "\n",
    "    # get signals within specific cell type\n",
    "    atac_peaks <- peaks[peaks$celltype == celltype,]\n",
    "    link_temp <- GRanges(atac_peaks[,c(\"seqnames\",\"peak.start\",\"peak.end\")]) #chr or seqnames\n",
    "\n",
    "    print(table(link_temp %in% anno_temp))\n",
    "    # OUTPUT filtered results\n",
    "    out <- atac_peaks[which(link_temp %in% anno_temp),]\n",
    "\n",
    "    final <- rbind(final, out)\n",
    "}\n",
    "table(final$celltype,useNA = \"always\")\n",
    "\n",
    "############################################# \n",
    "### Generate bed file for S-LDSC analysis ###\n",
    "#############################################\n",
    "############## for deg results ##############\n",
    "#############################################\n",
    "table(final$celltype)\n",
    "final <- final[order(c(final$seqnames)),]\n",
    "## generating peak region\n",
    "final$seqnames <- str_split_fixed(final$peak, \"-\",3)[,1]\n",
    "final$start <- str_split_fixed(final$peak, \"-\",3)[,2]\n",
    "final$end <- str_split_fixed(final$peak, \"-\",3)[,3]\n",
    "\n",
    "ct <- c(\"Astrocyte\",\"Excitatory\",\"Inhibitory\",\"Microglia\",\"Oligodendrocyte\",\"OPC\") #,\"OPC\"\n",
    "\n",
    "for (i in ct){\n",
    "    print(i)\n",
    "    df <- final[final$celltype == i,c(1,2,3)]\n",
    "    print(head(df))\n",
    "    write.table(\n",
    "        df, \n",
    "        file = paste(\"./Results/ldsc/DEG_linked_peaks_restricted/\",\"PFC_\",i,\"_DEG_linked_peaks_restricted.bed\",sep = \"\"),\n",
    "        quote=F,row.names=F, col.names=F, sep=\"\\t\")\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "##### restricting peaks called only in one cell type\n",
    "## OPTIONAL ##\n",
    "## generating cleaning cell type peaks files for six major cell types\n",
    "ct <- c(\"Astrocyte\",\"Excitatory\",\"Inhibitory\",\"Microglia\",\"Oligodendrocyte\",\"OPC\")\n",
    "# select specific peaks\n",
    "peaks <- ec_linked_peaks\n",
    "\n",
    "final <- data.frame()\n",
    "\n",
    "for(i in ct){\n",
    "    # assign cell type\n",
    "    celltype = i\n",
    "\n",
    "    ## ct peaks that are not cell type AD-DEG linked peaks\n",
    "    #anno_temp <- ctpeaks[grep(celltype,ctpeaks$peak_called_in)]\n",
    "    anno_temp <- ctpeaks[ctpeaks$peak_called_in == celltype]\n",
    "\n",
    "    # get signals within specific cell type\n",
    "    atac_peaks <- peaks[peaks$celltype == celltype,]\n",
    "    link_temp <- GRanges(atac_peaks[,c(\"seqnames\",\"peak.start\",\"peak.end\")]) #chr or seqnames\n",
    "\n",
    "    print(table(link_temp %in% anno_temp))\n",
    "    # OUTPUT filtered results\n",
    "    out <- atac_peaks[which(link_temp %in% anno_temp),]\n",
    "\n",
    "    final <- rbind(final, out)\n",
    "}\n",
    "table(final$celltype,useNA = \"always\")\n",
    "\n",
    "############################################# \n",
    "### Generate bed file for S-LDSC analysis ###\n",
    "#############################################\n",
    "############## for deg results ##############\n",
    "#############################################\n",
    "table(final$celltype)\n",
    "final <- final[order(c(final$seqnames)),]\n",
    "## generating peak region\n",
    "final$seqnames <- str_split_fixed(final$peak, \"-\",3)[,1]\n",
    "final$start <- str_split_fixed(final$peak, \"-\",3)[,2]\n",
    "final$end <- str_split_fixed(final$peak, \"-\",3)[,3]\n",
    "\n",
    "ct <- c(\"Astrocyte\",\"Excitatory\",\"Inhibitory\",\"Microglia\",\"Oligodendrocyte\",\"OPC\") #,\"OPC\"\n",
    "\n",
    "for (i in ct){\n",
    "    print(i)\n",
    "    df <- final[final$celltype == i,c(1,2,3)]\n",
    "    print(head(df))\n",
    "    write.table(\n",
    "        df, \n",
    "        file = paste(\"./Results/ldsc/DEG_linked_peaks_restricted/\",\"EC_\",i,\"_DEG_linked_peaks_restricted.bed\",sep = \"\"),\n",
    "        quote=F,row.names=F, col.names=F, sep=\"\\t\")\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "##### restricting peaks called only in one cell type\n",
    "## OPTIONAL ##\n",
    "## generating cleaning cell type peaks files for six major cell types\n",
    "ct <- c(\"Astrocyte\",\"Excitatory\",\"Inhibitory\",\"Microglia\",\"Oligodendrocyte\",\"OPC\")\n",
    "# select specific peaks\n",
    "peaks <- hip_linked_peaks\n",
    "\n",
    "final <- data.frame()\n",
    "\n",
    "for(i in ct){\n",
    "    # assign cell type\n",
    "    celltype = i\n",
    "\n",
    "    ## ct peaks that are not cell type AD-DEG linked peaks\n",
    "    #anno_temp <- ctpeaks[grep(celltype,ctpeaks$peak_called_in)]\n",
    "    anno_temp <- ctpeaks[ctpeaks$peak_called_in == celltype]\n",
    "\n",
    "    # get signals within specific cell type\n",
    "    atac_peaks <- peaks[peaks$celltype == celltype,]\n",
    "    link_temp <- GRanges(atac_peaks[,c(\"seqnames\",\"peak.start\",\"peak.end\")]) #chr or seqnames\n",
    "\n",
    "    print(table(link_temp %in% anno_temp))\n",
    "    # OUTPUT filtered results\n",
    "    out <- atac_peaks[which(link_temp %in% anno_temp),]\n",
    "\n",
    "    final <- rbind(final, out)\n",
    "}\n",
    "table(final$celltype,useNA = \"always\")\n",
    "\n",
    "############################################# \n",
    "### Generate bed file for S-LDSC analysis ###\n",
    "#############################################\n",
    "############## for deg results ##############\n",
    "#############################################\n",
    "table(final$celltype)\n",
    "final <- final[order(c(final$seqnames)),]\n",
    "## generating peak region\n",
    "final$seqnames <- str_split_fixed(final$peak, \"-\",3)[,1]\n",
    "final$start <- str_split_fixed(final$peak, \"-\",3)[,2]\n",
    "final$end <- str_split_fixed(final$peak, \"-\",3)[,3]\n",
    "\n",
    "ct <- c(\"Astrocyte\",\"Excitatory\",\"Inhibitory\",\"Microglia\",\"Oligodendrocyte\",\"OPC\") #,\"OPC\"\n",
    "\n",
    "for (i in ct){\n",
    "    print(i)\n",
    "    df <- final[final$celltype == i,c(1,2,3)]\n",
    "    print(head(df))\n",
    "    write.table(\n",
    "        df, \n",
    "        file = paste(\"./Results/ldsc/DEG_linked_peaks_restricted/\",\"HIP_\",i,\"_DEG_linked_peaks_restricted.bed\",sep = \"\"),\n",
    "        quote=F,row.names=F, col.names=F, sep=\"\\t\")\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "######################################################\n",
    "########### Working on S-LDSC results ################\n",
    "######################################################\n",
    "get_ldsc_results_deg <- function(dir = check,GWAS_oi = GWAS_oi){\n",
    "    GWAS_oi <- GWAS_oi\n",
    "    res_oi <- check[grep(GWAS_oi,check)]\n",
    "    res_oi <- res_oi[grep(\"DEG\",res_oi)]\n",
    "    res_deg <- data.frame()\n",
    "    for (i in res_oi){\n",
    "    filename <- paste(\"/data2/aliu8/2023_AD_multiome/Analysis/Results/ldsc/ldsc_out_deg_hg19/\", i, sep = \"\")\n",
    "    #print(filename)\n",
    "    df <- read.table(filename,header = 1)\n",
    "    df <- df[1,c(1,5:10)]\n",
    "    df$Category <- gsub(\".GWAS.results\",\"\",i)\n",
    "    df$Region <- str_split_fixed(df$Category,pattern = \"_\",4)[,3]\n",
    "    df$Category2 <- str_split_fixed(df$Category,pattern = \"_\",4)[,4]\n",
    "    df$Category3 <- str_split_fixed(df$Category2,pattern = \"_\",4)[,1]\n",
    "    res_deg <- rbind(res_deg, df)\n",
    "    }\n",
    "    #res_deg$Region <-rep(c(\"PFC\",\"EC\",\"HIP\"),6)\n",
    "    res_deg$Celltype <- res_deg$Category3\n",
    "    res_deg$cell_region <- paste(res_deg$Celltype,res_deg$Region,sep = \"_\")\n",
    "    res_deg$GWAS <- GWAS_oi\n",
    "    return(res_deg)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "########### Working on S-LDSC results ################\n",
    "######################################################\n",
    "## getting all file names\n",
    "check <- list.files(\"/data2/aliu8/2023_AD_multiome/Analysis/Results/ldsc/ldsc_out_deg_hg19/\")\n",
    "check <- check[grep(\".log\",check,invert = T)]\n",
    "#check\n",
    "trait_oi <- c(\"AD_Bellenguez\",\"AD_Kunkle\",\"AD_Wightman\",\"AD_Jansen\",\"AR_Watanabe\",\"Asthma_Watanabe\",\"Atopic_Watanabe\",\"CD_DeLange\",\"Celiac_Dubois\",\"Hypo_Watanabe\",\"MS_IMSGC2019\",\"PSC_Ji\",\n",
    "\"RA_Okada\",\"SLE_Bentham\",\"UC_DeLange\",\"Vitligo_Jin\",\"ADHD_Demontis\",\"ASD_Grove\",\"BIP_Mullins\",\"DEPRESSION_Howard\",\"MDD_Wray\",\"INSOMNIA_Schoeler\",\n",
    "\"NEUROTICISM_Nagel\",\"PTSD_Nievergelt\",\"SCZ_Trubetskoy\")\n",
    "\n",
    "## plotting\n",
    "res_deg <- data.frame()\n",
    "for (i in trait_oi){\n",
    "    GWASP_oi <- i\n",
    "    temp <- get_ldsc_results_deg(check, GWAS_oi = GWASP_oi)\n",
    "    res_deg <- rbind(res_deg, temp)\n",
    "}\n",
    "res_deg <- res_deg[order(res_deg$Category3),]\n",
    "#res_deg\n",
    "table(res_deg$GWAS)\n",
    "\n",
    "\n",
    "res_deg$p.adj <- NA\n",
    "for(i in trait_oi){\n",
    "    res_deg[res_deg$GWAS ==i,]$p.adj <- p.adjust(res_deg[res_deg$GWAS ==i,]$Enrichment_p,method = \"fdr\")\n",
    "}\n",
    "table(res_deg$Enrichment_p < 0.05)\n",
    "table(res_deg$p.adj < 0.05)\n",
    "write.csv(res_deg, file = \"./Results/ldsc/DEG_linked_peaks_hg19_ldsc.csv\",row.names = F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#### Figure 6d\n",
    "mat<-matrix(nrow=length(unique(res_deg$cell_region)), ncol=length(unique(res_deg$GWAS)))\n",
    "colnames(mat)<-unique(res_deg$GWAS)\n",
    "for (j in 1:length(unique(res_deg$cell_region))){\n",
    "  for (i in 1:length(unique(res_deg$GWAS))){\n",
    "    ctTmp<-unique(res_deg$cell_region)[j]\n",
    "    gwasTmp<-colnames(mat)[i]\n",
    "    sub<-res_deg[which(res_deg$cell_region==ctTmp & res_deg$GWAS==gwasTmp),]\n",
    "    if (nrow(sub)>0){\n",
    "      mat[j,i]<-sub$Coefficient_z.score\n",
    "    }\n",
    "    else{\n",
    "      mat[j,i]<-0\n",
    "    }\n",
    "  }\n",
    "}\n",
    "rownames(mat)<- unique(res_deg$cell_region)\n",
    "#mat\n",
    "sig_mat<-matrix(nrow=length(unique(res_deg$cell_region)), ncol=length(unique(res_deg$GWAS)))\n",
    "colnames(sig_mat)<-unique(res_deg$GWAS)\n",
    "for (j in 1:length(unique(res_deg$cell_region))){\n",
    "  for (i in 1:length(unique(res_deg$GWAS))){\n",
    "    ctTmp<-unique(res_deg$cell_region)[j]\n",
    "    gwasTmp<-colnames(mat)[i]\n",
    "    sub<-res_deg[which(res_deg$cell_region==ctTmp & res_deg$GWAS==gwasTmp),]\n",
    "    if (sub$Coefficient_z.score>0 & sub$p.adj <0.1){\n",
    "      sig_mat[j,i]<-sub$p.adj\n",
    "    }\n",
    "    else{\n",
    "      sig_mat[j,i]<-1\n",
    "    }\n",
    "  }\n",
    "}\n",
    "rownames(sig_mat)<- unique(res_deg$cell_region)\n",
    "#sig_mat\n",
    "\n",
    "library(ComplexHeatmap)\n",
    "library(circlize)\n",
    "library(RColorBrewer)\n",
    "ha<-rowAnnotation(Region=rownames(mat)\n",
    "                       , col= list(Region=c(\"Astrocyte_PFC\"=\"#f06719ff\",\"Astrocyte_EC\"=\"#f06719ff\",\"Astrocyte_HIP\"=\"#f06719ff\",\n",
    "                       \"Excitatory_PFC\"=\"#33a65cff\",\"Excitatory_EC\"=\"#33a65cff\",\"Excitatory_HIP\"=\"#33a65cff\",\n",
    "                       \"Inhibitory_PFC\"=\"#23767cff\",\"Inhibitory_EC\"=\"#23767cff\",\"Inhibitory_HIP\"=\"#23767cff\",\n",
    "                       \"Microglia_PFC\"=\"#e03426ff\",\"Microglia_EC\"=\"#e03426ff\",\"Microglia_HIP\"=\"#e03426ff\",\n",
    "                       \"Oligodendrocyte_PFC\"=\"#1ba3c6ff\",\"Oligodendrocyte_EC\"=\"#1ba3c6ff\",\"Oligodendrocyte_HIP\"=\"#1ba3c6ff\",\n",
    "                       \"OPC_PFC\"=\"#a26dc2ff\",\"OPC_EC\"=\"#a26dc2ff\",\"OPC_HIP\"=\"#a26dc2ff\")), show_legend=F,annotation_label=\"\")\n",
    "ha2<-rowAnnotation(Region=rownames(mat)\n",
    "                       , col= list(Region=c(\"Astrocyte_PFC\"=\"#c25757ff\",\"Astrocyte_EC\"=\"#825ca6ff\",\"Astrocyte_HIP\"=\"#3f78c1ff\",\n",
    "                       \"Excitatory_PFC\"=\"#c25757ff\",\"Excitatory_EC\"=\"#825ca6ff\",\"Excitatory_HIP\"=\"#3f78c1ff\",\n",
    "                       \"Inhibitory_PFC\"=\"#c25757ff\",\"Inhibitory_EC\"=\"#825ca6ff\",\"Inhibitory_HIP\"=\"#3f78c1ff\",\n",
    "                       \"Microglia_PFC\"=\"#c25757ff\",\"Microglia_EC\"=\"#825ca6ff\",\"Microglia_HIP\"=\"#3f78c1ff\",\n",
    "                       \"Oligodendrocyte_PFC\"=\"#c25757ff\",\"Oligodendrocyte_EC\"=\"#825ca6ff\",\"Oligodendrocyte_HIP\"=\"#3f78c1ff\",\n",
    "                       \"OPC_PFC\"=\"#c25757ff\",\"OPC_EC\"=\"#825ca6ff\",\"OPC_HIP\"=\"#3f78c1ff\")), show_legend=F,annotation_label=\"\")\n",
    "ha <- c(ha,ha2)\n",
    "ht=Heatmap(mat, cluster_rows=F,cluster_columns=F,\n",
    "           col=colorRamp2(seq(from = -4, to = 4,length = 11),\n",
    "           rev(RColorBrewer::brewer.pal(11,\"RdBu\"))),\n",
    "           show_column_names=T, show_row_names=F,row_names_side=\"left\",\n",
    "           column_names_side = \"top\",\n",
    "           left_annotation = ha,\n",
    "           row_names_max_width = max_text_width(rownames(mat),\n",
    "           gp = gpar(fontsize = 24)),\n",
    "           cell_fun = function(j, i, x, y, w, h, fill){\n",
    "            if(sig_mat[i, j] <0.01) {\n",
    "                    grid.text(\"**\", x, y, gp=gpar(fontsize=35, col=\"white\"), vjust=\"center\")\n",
    "                }else if (sig_mat[i, j] <0.05) {\n",
    "                   grid.text(\"*\", x, y, gp=gpar(fontsize=35, col=\"white\"), vjust=\"center\")\n",
    "                }\n",
    "            })\n",
    "pdf(\"/data2/aliu8/2023_AD_multiome/Analysis/Figures/LDSC_DEG_Linked_peaks_hg19.pdf\",height = 8,width = 16)\n",
    "ht\n",
    "dev.off()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
