{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## runnning Peak to gene links analysis on each brain regions\n",
    "# loading\n",
    "suppressPackageStartupMessages({\n",
    "library(Seurat)\n",
    "library(Signac)\n",
    "library(dplyr)\n",
    "\n",
    "library(RColorBrewer)\n",
    "library(ComplexHeatmap)\n",
    "library(circlize)\n",
    "library(stringr)\n",
    "library(GenomicRanges)\n",
    "library(GenomicFeatures)\n",
    "library(EnsDb.Hsapiens.v86)\n",
    "library(BSgenome.Hsapiens.UCSC.hg38)\n",
    "\n",
    "# library for plotting\n",
    "library(ggplot2)\n",
    "library(dplyr)\n",
    "library(viridis)\n",
    "})\n",
    "#### set working dir\n",
    "setwd(\"/data2/aliu8/2023_AD_multiome/Analysis/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Annotating the linked peaks\n",
    "## loading packages\n",
    "library(ChIPseeker)\n",
    "library(TxDb.Hsapiens.UCSC.hg38.knownGene)\n",
    "# downloaded hg38 known gene annotation from UCSC: https://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/genes/\n",
    "# read into txdb format\n",
    "txdb <- TxDb.Hsapiens.UCSC.hg38.knownGene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## build find link pipeline\n",
    "get_link <- function(object, celltype, dir, DEG_list){\n",
    "    \n",
    "    print(date())\n",
    "    \n",
    "    # print cell type\n",
    "    message(paste(\"Start working on \", celltype,\" in \",dir, \" direction.\",sep = \"\"))\n",
    "    \n",
    "    # subset object\n",
    "    #obj <- subset(object, subset = cluster_celltype == celltype)\n",
    "    obj <- object\n",
    "    print(obj)\n",
    "    print(table(obj$cluster_celltype))\n",
    "\n",
    "    # gene expression normalization\n",
    "    message(\"Noramlization in case not done before.\")\n",
    "    DefaultAssay(obj) <- \"PC\"\n",
    "    obj <- SCTransform(obj,assay = \"PC\")\n",
    "\n",
    "    #subset degs\n",
    "    DEGs <- unique(DEG_list[DEG_list$dir == dir & DEG_list$celltype == celltype,]$gene)\n",
    "    message(length(DEGs))\n",
    "\n",
    "    # atac set normalization \n",
    "    DefaultAssay(obj) <- \"CTpeaks\"\n",
    "    obj <- RunTFIDF(obj) %>% \n",
    "                FindTopFeatures(min.cutoff='q0') %>% \n",
    "                RunSVD()\n",
    "    \n",
    "    # first compute the GC content for each peak\n",
    "    obj <- RegionStats(obj, genome = BSgenome.Hsapiens.UCSC.hg38)\n",
    "\n",
    "    # link peaks to genes\n",
    "    message(\"Start calculating links.\")\n",
    "    \n",
    "    obj <- LinkPeaks(\n",
    "        object = obj,\n",
    "        peak.assay = \"CTpeaks\",\n",
    "        peak.slot = \"data\",\n",
    "        expression.assay = \"SCT\",\n",
    "        expression.slot = \"data\",\n",
    "        genes.use = DEGs,\n",
    "        method = \"spearman\",#c(\"pearson\",\"spearman\") # spearman may preferred.\n",
    "        score_cutoff = 0.05, # set score to 0 to keep all \n",
    "        pvalue_cutoff = 0.05 # set p val = 1 to include all independent tests\n",
    "    )\n",
    "\n",
    "    # clean results and calculate adjust pvalues\n",
    "    ## preform multiple testing correction\n",
    "    message(\"Final processing the results.\")\n",
    "    res <- as.data.frame(Links(obj))\n",
    "    res$p.adj <- p.adjust(res$pvalue,method = \"BH\")\n",
    "    res <- res[res$score > 0.05 & res$p.adj < 0.05,]\n",
    "\n",
    "    ## add annotataion information\n",
    "    res$celltype <- celltype\n",
    "    res$deg_dir <- dir\n",
    "    res$comb <- paste(res$peak,res$gene,res$celltype,res$deg_dir, sep = \"_\")\n",
    "    message(paste(\"Identified\",length(res$comb),\"links.\"))\n",
    "\n",
    "    # reture\n",
    "    return(res)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# pipeline to keep peaks only expressed in assigned celltype.  \n",
    "get_clean_link <- function(links_df,ctpeaks){\n",
    "    linked_peaks <- links_df\n",
    "\n",
    "    # reformat the peak regions \n",
    "    linked_peaks$peak.start <-  str_split_fixed(linked_peaks$peak,\"-\",3)[,2]\n",
    "    linked_peaks$peak.end <- str_split_fixed(linked_peaks$peak,\"-\",3)[,3]\n",
    "    linked_peaks <- unique(linked_peaks)\n",
    "    table(linked_peaks$celltype)\n",
    "\n",
    "    final <- data.frame()\n",
    "\n",
    "    ct <- c(\"Astrocyte\",\"Excitatory\",\"Inhibitory\",\"Microglia\",\"Oligodendrocyte\",\"OPC\")\n",
    "    \n",
    "    for(i in ct){\n",
    "        # assign cell type\n",
    "        celltype = i\n",
    "\n",
    "        ## ct peaks that are not cell type AD-DEG linked peaks\n",
    "        anno_temp <- ctpeaks[grep(celltype,ctpeaks$peak_called_in)]\n",
    "\n",
    "        # get signals within specific cell type\n",
    "        atac_peaks <- linked_peaks[linked_peaks$celltype == celltype,]\n",
    "\n",
    "        link_temp <- GRanges(atac_peaks[,c(\"seqnames\",\"peak.start\",\"peak.end\")])\n",
    "\n",
    "        print(table(link_temp %in% anno_temp))\n",
    "        # OUTPUT filtered results\n",
    "        out <- atac_peaks[which(link_temp %in% anno_temp),]\n",
    "\n",
    "        final <- rbind(final, out)\n",
    "    }\n",
    "\n",
    "    return(final)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## CT peaks annotation files by MACS2\n",
    "ctpeaks <- read.csv(\"./Results/CTpeaks_annotated.csv\",row.names = 1)\n",
    "colnames(ctpeaks) <- c(\"chr\",\"start\",\"end\",\"width\",\"strand\",\"peak_called_in\")\n",
    "ctpeaks <- GRanges(ctpeaks)\n",
    "ctpeaks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "### PFC ###\n",
    "## read in related data \n",
    "object <- readRDS(\"cellbender_PFC_object.rds\")\n",
    "\n",
    "## read in DEG list\n",
    "DEG_list <- read.csv(\"./Results/DEG/Overlap_mast_mixed_PFC.csv\")\n",
    "\n",
    "## actuall link analysis\n",
    "ct <- c(\"Astrocyte\",\"Excitatory\",\"Inhibitory\",\"Microglia\",\"Oligodendrocyte\",\"OPC\")\n",
    "\n",
    "direction <- c(\"pos\",\"neg\")\n",
    "\n",
    "## running\n",
    "final <- data.frame()\n",
    "for (i in ct){\n",
    "    for(j in direction){\n",
    "        celltype <- i\n",
    "        print(celltype)\n",
    "        dir = j\n",
    "        print(dir)\n",
    "        \n",
    "        res1 <- get_link(object=object, celltype=celltype, dir=dir, DEG_list=DEG_list)\n",
    "        print(length(res1$gene))\n",
    "        final <- rbind(final,res1)\n",
    "    }\n",
    "}\n",
    "\n",
    "# get peaks only shown in assigned celltype\n",
    "#table(final$celltype, final$deg_dir)\n",
    "links_clean <- get_clean_link(links_df = final, ctpeaks = ctpeaks)\n",
    "\n",
    "## check information\n",
    "table(links_clean$celltype,links_clean$deg_dir)\n",
    "\n",
    "## saving the peak-gene file\n",
    "write.csv(links_clean, file = \"./Results/LINK/PFC_linkpeaks_all.csv\",row.names = F)\n",
    "\n",
    "### Figure 3c: PFC\n",
    "# histogram plotting for linkage between peaks and genes in each brain region\n",
    "df <- table(paste(links_clean$gene,links_clean$celltype,sep = \"_\"))\n",
    "#hist(df)\n",
    "df <- as.data.frame(df)\n",
    "\n",
    "avg.peaks <- round(median(df[,2]),digits = 0)\n",
    "message(paste(\"Mean of number of peaks linked to DEGs: \",avg.peaks,\".\",sep = \"\"))\n",
    "\n",
    "\n",
    "df$Freq <- ifelse(df$Freq >39,40,df$Freq)\n",
    "# Represent it\n",
    "p1 <- df %>%\n",
    "  ggplot( aes(x=Freq)) +\n",
    "    geom_histogram(fill=\"#c25757ff\", alpha=0.7, position = 'dodge',binwidth = 2,colour='#c25757ff',size=1) +\n",
    "    scale_fill_viridis(discrete=TRUE)+\n",
    "    xlab(\"Linked peaks per gene\")+\n",
    "    ylab(\"Number of genes\")+\n",
    "    labs(fill=\"PFC\")+theme_classic()+ggtitle(paste(\"PFC: \",avg.peaks,\"linked peaks per DEG.\"))\n",
    "\n",
    "pdf(file = \"./Figures/LINK/Hist_link.peaks_PFC.pdf\",width = 5,height = 5)\n",
    "p1\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "### EC ###\n",
    "## read in related data\n",
    "object <- readRDS(\"cellbender_EC_object.rds\")\n",
    "\n",
    "## read in DEG list\n",
    "DEG_list <- read.csv(\"./Results/DEG/Overlap_mast_mixed_EC.csv\")\n",
    "\n",
    "## actuall link analysis\n",
    "ct <- c(\"Astrocyte\",\"Excitatory\",\"Inhibitory\",\"Microglia\",\"Oligodendrocyte\",\"OPC\")\n",
    "\n",
    "direction <- c(\"pos\",\"neg\")\n",
    "\n",
    "## running\n",
    "final <- data.frame()\n",
    "for (i in ct){\n",
    "    for(j in direction){\n",
    "        celltype <- i\n",
    "        print(celltype)\n",
    "        dir = j\n",
    "        print(dir)\n",
    "        \n",
    "        res1 <- get_link(object=object, celltype=celltype, dir=dir, DEG_list=DEG_list)\n",
    "        print(length(res1$gene))\n",
    "        final <- rbind(final,res1)\n",
    "    }\n",
    "}\n",
    "\n",
    "# get peaks only shown in assigned celltype\n",
    "#table(final$celltype, final$deg_dir)\n",
    "links_clean <- get_clean_link(links_df = final, ctpeaks = ctpeaks)\n",
    "\n",
    "## check information\n",
    "table(links_clean$celltype,links_clean$deg_dir)\n",
    "\n",
    "## saving the peak-gene file\n",
    "write.csv(links_clean, file = \"./Results/LINK/EC_linkpeaks_all.csv\",row.names = F)\n",
    "\n",
    "### Figure 3c: EC\n",
    "# histogram plotting for linkage between peaks and genes in each brain region\n",
    "df <- table(paste(links_clean$gene,links_clean$celltype,sep = \"_\"))\n",
    "#hist(df)\n",
    "df <- as.data.frame(df)\n",
    "\n",
    "avg.peaks <- round(median(df[,2]),digits = 0)\n",
    "message(paste(\"Mean of number of peaks linked to DEGs: \",avg.peaks,\".\",sep = \"\"))\n",
    "\n",
    "\n",
    "df$Freq <- ifelse(df$Freq >39,40,df$Freq)\n",
    "# Represent it\n",
    "p1 <- df %>%\n",
    "  ggplot( aes(x=Freq)) +\n",
    "    geom_histogram(fill=\"#825ca6ff\", alpha=0.7, position = 'dodge',binwidth = 2,colour='#c25757ff',size=1) +\n",
    "    scale_fill_viridis(discrete=TRUE)+\n",
    "    xlab(\"Linked peaks per gene\")+\n",
    "    ylab(\"Number of genes\")+\n",
    "    labs(fill=\"EC\")+theme_classic()+ggtitle(paste(\"EC: \",avg.peaks,\"linked peaks per DEG.\"))\n",
    "\n",
    "pdf(file = \"./Figures/LINK/Hist_link.peaks_EC.pdf\",width = 5,height = 5)\n",
    "p1\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "### HIP ###\n",
    "## read in related data\n",
    "object <- readRDS(\"cellbender_HIP_object.rds\")\n",
    "\n",
    "## read in DEG list\n",
    "DEG_list <- read.csv(\"./Results/DEG/Overlap_mast_mixed_HIP.csv\")\n",
    "\n",
    "## actuall link analysis\n",
    "ct <- c(\"Astrocyte\",\"Excitatory\",\"Inhibitory\",\"Microglia\",\"Oligodendrocyte\",\"OPC\")\n",
    "\n",
    "direction <- c(\"pos\",\"neg\")\n",
    "\n",
    "## running\n",
    "final <- data.frame()\n",
    "for (i in ct){\n",
    "    for(j in direction){\n",
    "        celltype <- i\n",
    "        print(celltype)\n",
    "        dir = j\n",
    "        print(dir)\n",
    "        \n",
    "        res1 <- get_link(object=object, celltype=celltype, dir=dir, DEG_list=DEG_list)\n",
    "        print(length(res1$gene))\n",
    "        final <- rbind(final,res1)\n",
    "    }\n",
    "}\n",
    "\n",
    "# get peaks only shown in assigned celltype\n",
    "#table(final$celltype, final$deg_dir)\n",
    "links_clean <- get_clean_link(links_df = final, ctpeaks = ctpeaks)\n",
    "\n",
    "## check information\n",
    "table(links_clean$celltype,links_clean$deg_dir)\n",
    "\n",
    "## saving the peak-gene file\n",
    "write.csv(links_clean, file = \"./Results/LINK/HIP_linkpeaks_all.csv\",row.names = F)\n",
    "\n",
    "### Figure 3c: HIP\n",
    "# histogram plotting for linkage between peaks and genes in each brain region\n",
    "df <- table(paste(links_clean$gene,links_clean$celltype,sep = \"_\"))\n",
    "#hist(df)\n",
    "df <- as.data.frame(df)\n",
    "\n",
    "avg.peaks <- round(median(df[,2]),digits = 0)\n",
    "message(paste(\"Mean of number of peaks linked to DEGs: \",avg.peaks,\".\",sep = \"\"))\n",
    "\n",
    "\n",
    "df$Freq <- ifelse(df$Freq >39,40,df$Freq)\n",
    "# Represent it\n",
    "p1 <- df %>%\n",
    "  ggplot( aes(x=Freq)) +\n",
    "    geom_histogram(fill=\"#c25757ff\", alpha=0.7, position = 'dodge',binwidth = 2,colour='#c25757ff',size=1) +\n",
    "    scale_fill_viridis(discrete=TRUE)+\n",
    "    xlab(\"Linked peaks per gene\")+\n",
    "    ylab(\"Number of genes\")+\n",
    "    labs(fill=\"HIP\")+theme_classic()+ggtitle(paste(\"HIP: \",avg.peaks,\"linked peaks per DEG.\"))\n",
    "\n",
    "pdf(file = \"./Figures/LINK/Hist_link.peaks_HIP.pdf\",width = 5,height = 5)\n",
    "p1\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "### get annotation of the peaks linked to EOAD-DEG ##\n",
    "## reference peaks bed Li et al. 2023 A comparative atlas of single-cell chrommatin accessibitlity in the human brain \n",
    "brain.peaks <- read.delim(\"./Data/cCREs.bed\", header = F)\n",
    "colnames(brain.peaks) <- c(\"chr\",\"start\",\"end\",\"annotation\")\n",
    "brain.peaks <- GRanges(brain.peaks)\n",
    "head(brain.peaks)\n",
    "\n",
    "## ENCODE cCRE peaks #download from https://genome.ucsc.edu/cgi-bin/hgTrackUi?db=hg38&g=encodeCcreCombined\n",
    "encode_peaks <- import.bb(\"./Data/encodeCcreCombined.bb\")\n",
    "head(encode_peaks)\n",
    "\n",
    "## add brain single cell eqtl information to the linked peaks\n",
    "## get eqtl information\n",
    "brain_eqlt <- read.csv(\"/data2/aliu8/2023_AD_multiome/Analysis/Data/brain_sc_eqtl_sig.csv\")\n",
    "brain_eqlt <-  brain_eqlt[!is.na(brain_eqlt$chr),]\n",
    "brain_eqlt$chr <- paste(\"chr\",brain_eqlt$chr,sep = \"\")\n",
    "dim(brain_eqlt)\n",
    "#head(brain_eqlt)\n",
    "\n",
    "# from expanded snps list, generate a GRanges format file\n",
    "#snp_coords <- as.data.frame(sig[,c(\"chr\",\"variant_pos\",\"variant_pos\",\"rs_id_dbSNP151_GRCh38p7\",\"gene_name\")])\n",
    "snp_coords <- as.data.frame(brain_eqlt[,c(\"chr\",\"pos\",\"pos\",\"SNP\",\"symbol\",\"cell_type\",\"dist_TSS\",\"beta\")])\n",
    "\n",
    "colnames(snp_coords) <- c(\"chr\",\"start\",\"end\",\"RSID\",\"gene_name\",\"cell_type\",\"dist_TSS\",\"beta\")\n",
    "snp_coords <- GRanges(snp_coords)\n",
    "head(snp_coords)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## create function for peak annotation\n",
    "get_peak_annotation <- function(linked_peaks = linked_peaks, encode_peaks = encode_peaks, brain.peaks = brain.peaks){\n",
    "    linked_peaks = linked_peaks\n",
    "    ## change the format of linked peaks to GRanges\n",
    "    #linked_peaks$peak.start <-  str_split_fixed(linked_peaks$peak,\"-\",3)[,2]\n",
    "    #linked_peaks$peak.end <- str_split_fixed(linked_peaks$peak,\"-\",3)[,3]\n",
    "    linked_gr <- linked_peaks[,c(1,14,15,4:13)]\n",
    "    #linked_gr <- linked_peaks[,c(1,14,15,4:13,16)]\n",
    "    df_granges <- GRanges(linked_gr)\n",
    "\n",
    "    ## find overlap with encode and brain cCRE\n",
    "    overlap_encode <- findOverlaps(df_granges,encode_peaks)\n",
    "    overlap_brain_ccre <- findOverlaps(df_granges,brain.peaks)\n",
    "    overlap_eqtl <- findOverlaps(df_granges,snp_coords)\n",
    "\n",
    "    ## show results\n",
    "    print(\"EOAD-DEG linked peaks overlap with ENCODE cis-CRE\")\n",
    "    print(length(unique(queryHits(overlap_encode))))\n",
    "    print(length(unique(queryHits(overlap_encode)))/length(df_granges))\n",
    "\n",
    "    print(\"EOAD-DEG linked peaks overlap with brain cis-CRE atlas\")\n",
    "    print(length(unique(queryHits(overlap_brain_ccre))))\n",
    "    print(length(unique(queryHits(overlap_brain_ccre)))/length(df_granges))\n",
    "\n",
    "    print(\"EOAD-DEG linked peaks overlap with single cell brain eqtl\")\n",
    "    # ATAC-seq peaks contain SNPs\n",
    "    peaks_contain_snps <- linked_peaks[queryHits(overlap_eqtl),]\n",
    "    peaks_contain_snps$RSID <- snp_coords[subjectHits(overlap_eqtl)]$RSID\n",
    "    peaks_contain_snps$eGene <- snp_coords[subjectHits(overlap_eqtl)]$gene_name\n",
    "    peaks_contain_snps$cell_type <- snp_coords[subjectHits(overlap_eqtl)]$cell_type\n",
    "    peaks_contain_snps$dist_TSS <- snp_coords[subjectHits(overlap_eqtl)]$dist_TSS\n",
    "    peaks_contain_snps$beta <- snp_coords[subjectHits(overlap_eqtl)]$beta\n",
    "    peaks_contain_snps <- peaks_contain_snps[peaks_contain_snps$gene == peaks_contain_snps$eGene & peaks_contain_snps$celltype == peaks_contain_snps$cell_type,]\n",
    "\n",
    "    print(length(unique(rownames(peaks_contain_snps))))\n",
    "    print(length(unique(rownames(peaks_contain_snps)))/length(df_granges))\n",
    "\n",
    "    ## add annotation if in ENCODE or brain cCRE atlas\n",
    "    linked_peaks$in_encode <- FALSE\n",
    "    linked_peaks[unique(queryHits(overlap_encode)),]$in_encode <- TRUE\n",
    "    linked_peaks$in_brain_cCRE <- FALSE\n",
    "    linked_peaks[unique(queryHits(overlap_brain_ccre)),]$in_brain_cCRE <- TRUE\n",
    "    linked_peaks$in_brain_sc_eqlt <- ifelse(linked_peaks$comb %in% peaks_contain_snps$comb, T, F)\n",
    "\n",
    "    ## get annotation from ENCODE\n",
    "    q_df <- linked_peaks[queryHits(overlap_encode),]\n",
    "    s_df <- as.data.frame(encode_peaks[subjectHits(overlap_encode)])\n",
    "\n",
    "    # combine information \n",
    "    c_df <- cbind(q_df,s_df)\n",
    "    c_df$encodeLabel <- factor(c_df$encodeLabel,levels = c(\"PLS\",\"pELS\",\"DNase-H3K4me3\",\"dELS\",\"CTCF-only\",\"Other\"))\n",
    "    c_df <- c_df[order(c_df$encodeLabel,c_df$celltype,c_df$deg_dir,c_df$seqnames,c_df$start),]\n",
    "\n",
    "    # based on the priority, keep the first one\n",
    "    f_df <- c_df[!duplicated(c_df$comb),]\n",
    "\n",
    "    ### add annotation information to original data\n",
    "    id <- match(linked_peaks$comb,f_df$comb)\n",
    "\n",
    "    linked_peaks$encodeLabel <- f_df[id,]$encodeLabel\n",
    "    linked_peaks$ucscLabel <- f_df[id,]$ucscLabel\n",
    "    linked_peaks$description <- f_df[id,]$description\n",
    "    ## change na to other\n",
    "    linked_peaks[is.na(linked_peaks$encodeLabel),]$encodeLabel <- \"Other\"\n",
    "    linked_peaks[is.na(linked_peaks$ucscLabel),]$ucscLabel <- \"Other\"\n",
    "\n",
    "    ## adding RSID information for brain eqtl\n",
    "    # add rsid information for egene\n",
    "    linked_peaks$eqtl_RSID <- NA\n",
    "    linked_peaks$dist_TSS <- NA\n",
    "    linked_peaks$beta <- NA\n",
    "    id <- match(peaks_contain_snps$comb,linked_peaks$comb)\n",
    "    linked_peaks[id,]$eqtl_RSID <- peaks_contain_snps$RSID\n",
    "    linked_peaks[id,]$dist_TSS <- peaks_contain_snps$dist_TSS\n",
    "    linked_peaks[id,]$beta <- peaks_contain_snps$beta\n",
    "\n",
    "    return(linked_peaks)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "### peak annotation on each region\n",
    "## PFC\n",
    "pfc_linked_peaks <- read.csv(\"./Results/LINK/PFC_linkpeaks_all.csv\",header = 1)\n",
    "pfc_linked_peaks <- pfc_linked_peaks[,-11]\n",
    "pfc_linked_peaks$p.adj <- p.adjust(pfc_linked_peaks$pvalue,method = \"BH\")\n",
    "table(pfc_linked_peaks$p.adj < 0.05)\n",
    "pfc_linked_peaks_annotated <- get_peak_annotation(linked_peaks = pfc_linked_peaks,encode_peaks = encode_peaks, brain.peaks = brain.peaks)\n",
    "#pfc_linked_peaks_annotated[pfc_linked_peaks_annotated$in_brain_sc_eqlt,]\n",
    "write.csv(pfc_linked_peaks_annotated, file = \"./Results/LINK/PFC_linkpeaks_all_annotated.csv\", row.names = F)\n",
    "\n",
    "## EC\n",
    "ec_linked_peaks <- read.csv(\"./Results/LINK/EC_linkpeaks_all.csv\",header = 1)\n",
    "ec_linked_peaks$p.adj <- p.adjust(ec_linked_peaks$pvalue,method = \"BH\")\n",
    "table(ec_linked_peaks$p.adj < 0.05)\n",
    "ec_linked_peaks_annotated <- get_peak_annotation(linked_peaks = ec_linked_peaks,encode_peaks = encode_peaks, brain.peaks = brain.peaks)\n",
    "#ec_linked_peaks_annotated[ec_linked_peaks_annotated$in_brain_sc_eqlt,]\n",
    "write.csv(ec_linked_peaks_annotated, file = \"./Results/LINK/EC_linkpeaks_all_annotated.csv\", row.names = F)\n",
    "\n",
    "## HIP\n",
    "hip_linked_peaks <- read.csv(\"./Results/LINK/HIP_linkpeaks_all.csv\",header = 1)\n",
    "hip_linked_peaks$p.adj <- p.adjust(hip_linked_peaks$pvalue,method = \"BH\")\n",
    "table(hip_linked_peaks$p.adj < 0.05)\n",
    "hip_linked_peaks_annotated <- get_peak_annotation(linked_peaks = hip_linked_peaks,encode_peaks = encode_peaks, brain.peaks = brain.peaks)\n",
    "#hip_linked_peaks_annotated[hip_linked_peaks_annotated$in_brain_sc_eqlt,]\n",
    "write.csv(hip_linked_peaks_annotated, file = \"./Results/LINK/HIP_linkpeaks_all_annotated.csv\", row.names = F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "### check annotated results ###\n",
    "pfc_linked_peaks_annotated <- read.csv(\"./Results/LINK/PFC_linkpeaks_all_annotated.csv\")\n",
    "ec_linked_peaks_annotated <- read.csv(\"./Results/LINK/EC_linkpeaks_all_annotated.csv\")\n",
    "hip_linked_peaks_annotated <- read.csv(\"./Results/LINK/HIP_linkpeaks_all_annotated.csv\")\n",
    "\n",
    "table(pfc_linked_peaks_annotated$celltype, pfc_linked_peaks_annotated$deg_dir)\n",
    "table(ec_linked_peaks_annotated$celltype, ec_linked_peaks_annotated$deg_dir)\n",
    "table(hip_linked_peaks_annotated$celltype, hip_linked_peaks_annotated$deg_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "### Figure 3d.\n",
    "## draw proportional plot of ENCODE cCRE annotation\n",
    "pfc_pro <- prop.table(table(pfc_linked_peaks_annotated$encodeLabel))\n",
    "ec_pro <- prop.table(table(ec_linked_peaks_annotated$encodeLabel))\n",
    "hip_pro <- prop.table(table(hip_linked_peaks_annotated$encodeLabel))\n",
    "\n",
    "pfc_pro <- as.data.frame(pfc_pro)\n",
    "pfc_pro$region <- 'PFC'\n",
    "ec_pro <- as.data.frame(ec_pro)\n",
    "ec_pro$region <- 'EC'\n",
    "hip_pro <- as.data.frame(hip_pro)\n",
    "hip_pro$region <- 'HIP'\n",
    "\n",
    "\n",
    "\n",
    "temp <- rbind(pfc_pro,ec_pro,hip_pro)\n",
    "temp$Region <- factor(temp$region,levels = c(\"HIP\",\"EC\",\"PFC\"))\n",
    "colnames(temp)[1] <- \"Annotation\"\n",
    "\n",
    "pdf(\"/data2/aliu8/2023_AD_multiome/Analysis/Figures/LINK/region_linked_peaks_annotation_3.29.pdf\",width = 12,height = 3)    \n",
    "ggplot(temp, aes(x=Freq, y=Region, fill=Annotation)) + \n",
    "    geom_bar(stat=\"identity\")+xlab(\"Proportion of Linked peaks\") + \n",
    "    ylab(\"Region\") + \n",
    "    theme_classic() +\n",
    "    scale_fill_manual(values=c(\"#FF0000\",\"#FFA700\",\"#FFAAAA\",\"#FFCD00\",\"#00B0F0\",\"#706f6f\"),labels=c(\"PLS\",\"pELS\",\"DNase-H3K4me3\",\"dELS\",\"CTCF-only\",\"Other\")) +\n",
    "    theme(legend.position=\"right\", axis.text=element_text(size=12),axis.title=element_text(size=15), legend.text=element_text(size=12))\n",
    "        \n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "### Supplementary Figs ENCODE cCREs annotation on each regions\n",
    "#linked_peaks_annotated <- ec_linked_peaks_annotated\n",
    "linked_peaks_annotated <- pfc_linked_peaks_annotated\n",
    "#linked_peaks_annotated <- hip_linked_peaks_annotated\n",
    "## create a dataframe of proportion\n",
    "df <- data.frame()\n",
    "ct <- c('Astrocyte','Excitatory','Inhibitory','Microglia','Oligodendrocyte','OPC')\n",
    "for (i in 1:length(ct)){\n",
    "    celltype <- ct[i]\n",
    "    temp <- as.data.frame(prop.table(table(linked_peaks_annotated[linked_peaks_annotated$celltype == celltype,]$encodeLabel)))\n",
    "    temp$celltype <- celltype\n",
    "\n",
    "    df <- rbind(df, temp)\n",
    "}\n",
    "\n",
    "colnames(df)[1] <- \"Annotation\"\n",
    "pdf(\"./Figures/LINK/region_linked_peaks_annotation_PFC.pdf\",width = 10,height = 4)    \n",
    "ggplot(df, aes(x=Freq, y=celltype, fill=Annotation)) + \n",
    "    geom_bar(stat=\"identity\")+xlab(\"Proportion of Linked peaks\") + \n",
    "    ylab(\"Cell type\") + \n",
    "    theme_classic() +\n",
    "    scale_fill_manual(values=c(\"#FF0000\",\"#FFA700\",\"#FFAAAA\",\"#FFCD00\",\"#00B0F0\",\"#706f6f\"),labels=c(\"PLS\",\"pELS\",\"DNase-H3K4me3\",\"dELS\",\"CTCF-only\",\"Other\")) +\n",
    "    theme(legend.position=\"right\", axis.text=element_text(size=12),axis.title=element_text(size=15), legend.text=element_text(size=12))\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "### Motif enrichment analysis\n",
    "## loading object\n",
    "pfc <- readRDS(\"cellbender_PFC_object.rds\")\n",
    "ec <- readRDS(\"cellbender_EC_object.rds\")\n",
    "hip <- readRDS(\"cellbender_HIP_object.rds\")\n",
    "\n",
    "## download human CORE TFBS from JASPAR\n",
    "fn <- file.path(\"/data2/aliu8/2023_AD_multiome/Analysis/Data/JASPAR2024_hs_core_755.txt\")\n",
    "fn\n",
    "\n",
    "pfm <- readJASPARMatrix(fn, matrixClass=\"PFM\")\n",
    "#names(pfm) <- str_split_fixed(names(pfm),\"\\\\.\",n = 3)[,3]\n",
    "pfm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# build function to run motif enrichment by cell type\n",
    "get_motif <- function(object,celltype,deg_direction){\n",
    "    # get interest peaks\n",
    "    peak_interest <- links[links$celltype == celltype & links$deg_dir == deg_direction & links$score >0.05,]$peak\n",
    "    message(paste(\"Testing on\",length(peak_interest),\"peaks links to\",celltype,deg_direction,\"DEGs.\"))\n",
    "\n",
    "    # find peaks open in selected cell type\n",
    "    DefaultAssay(object) <- \"CTpeaks\"\n",
    "    open.peaks <- AccessiblePeaks(object, idents = celltype)\n",
    "    \n",
    "    # match the overall GC content in the peak set\n",
    "    meta.feature <- GetAssayData(object, assay = \"CTpeaks\", slot = \"meta.features\")\n",
    "    peaks.matched <- MatchRegionStats(\n",
    "        meta.feature = meta.feature[open.peaks, ],\n",
    "        query.feature = meta.feature[peak_interest, ],\n",
    "        n = 50000)\n",
    "\n",
    "    ## test enrichment\n",
    "    enriched.motifs <- FindMotifs(\n",
    "        object = object,\n",
    "        features = peak_interest,\n",
    "        background=peaks.matched)\n",
    "    \n",
    "    enriched.motifs$celltype <- celltype\n",
    "    enriched.motifs$deg_dir <- deg_direction\n",
    "    \n",
    "    # getting significant results\n",
    "    sig.enriched.motifs <- enriched.motifs[enriched.motifs$p.adjust < 0.05,]\n",
    "    \n",
    "    # results\n",
    "    return(sig.enriched.motifs)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## Get mortif information for finding TF - peaks - differential correlation - DEGs\n",
    "motif.all <- GetMotifData(\n",
    "    object = object, assay = \"CTpeaks\", slot = \"data\"\n",
    "  )\n",
    "# motif.all[1,]\n",
    "motif.names <- GetMotifData(\n",
    "    object = object, assay = \"CTpeaks\", slot = \"motif.names\"\n",
    "  )\n",
    "length(motif.names)\n",
    "#names(motif.names$V1) <- str_split_fixed(names(motif.names),\"\\\\.\",n = 3)[,3]\n",
    "# generate dataframe of motif name and presents.\n",
    "motif.names <- as.data.frame(as.matrix(motif.names))\n",
    "motif.names$V1 <- str_split_fixed(motif.names$V1,\"\\\\.\",n = 3)[,3]\n",
    "motif.names$V2 <- str_split_fixed(motif.names$V1,\"::\",n = 2)[,1]\n",
    "\n",
    "head(motif.names)\n",
    "#unlist(motif.names)\n",
    "#rownames(motif.names[motif.names$V1 == \"ASCL1\",])\n",
    "table(unique(motif.names$V1) %in% rownames(object@assays$PC))\n",
    "table(unique(motif.names$V2) %in% rownames(object@assays$PC))\n",
    "\n",
    "length(unique(motif.names$V2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## running analysis on PFC region object\n",
    "# load data\n",
    "object <- pfc\n",
    "object\n",
    "\n",
    "## top upregulated genes\n",
    "degs <- read.csv(\"./Results/DEG/Overlap_mast_mixed_PFC.csv\",row.names = 1)\n",
    "message(\"How many DEGs in each cell type in PFC.\")\n",
    "dim(degs)\n",
    "table(degs$celltype,degs$dir)\n",
    "#degs\n",
    "\n",
    "## read in the link data\n",
    "links<- read.csv(\"./Results/LINK/PFC_linkpeaks_all_annotated.csv\")\n",
    "links$new_comb <- paste(links$gene,links$deg_dir,links$celltype,sep = \"_\")\n",
    "#\n",
    "\n",
    "message(\"How many linked peaks in each cell type in PFC.\")\n",
    "dim(links)\n",
    "table(links$celltype,links$deg_dir)\n",
    "#\n",
    "\n",
    "message(\"How many linked DEGs in each cell type in PFC.\")\n",
    "degs <- degs[degs$comb %in% unique(links$new_comb),]\n",
    "table(degs$celltype,degs$dir)\n",
    "length(degs$comb)\n",
    "#degs\n",
    "links <- links[links$new_comb %in% degs$comb,]\n",
    "\n",
    "### running analysis\n",
    "ct <- c(\"Astrocyte\",\"Excitatory\",\"Inhibitory\",\"Microglia\",\"Oligodendrocyte\",\"OPC\")\n",
    "direction <- c(\"pos\",\"neg\")\n",
    "Idents(object) <- \"cluster_celltype\"\n",
    "final <- data.frame()\n",
    "\n",
    "## running\n",
    "for (i in ct){\n",
    "    for(j in direction){\n",
    "        celltype <- i\n",
    "        print(celltype)\n",
    "        deg_direction = j\n",
    "        print(dir)\n",
    "        \n",
    "        res1 <- get_motif(object=object, celltype=celltype, deg_direction = deg_direction)\n",
    "        print(length(res1$motif.name))\n",
    "        final <- rbind(final,res1)\n",
    "    }\n",
    "}\n",
    "\n",
    "#### generate heatmap for motif enriched results_PFC\n",
    "enriched.motif <- final\n",
    "table(enriched.motif$celltype,enriched.motif$deg_dir)\n",
    "\n",
    "enriched.motif$motif.name <- str_split_fixed(enriched.motif$motif.name,\"\\\\.\",n = 3)[,3]\n",
    "enriched.motif$motif.name <- str_split_fixed(enriched.motif$motif.name,\"::\",n = 1)[,1]\n",
    "\n",
    "head(enriched.motif)\n",
    "\n",
    "## only keeping those TFs expressed in the specific cell type. \n",
    "expr.res <- data.frame()\n",
    "ct = c(\"Astrocyte\",\"Excitatory\",\"Inhibitory\",\"Microglia\",\"Oligodendrocyte\",\"OPC\")\n",
    "\n",
    "for (i in ct){\n",
    "    print(i)\n",
    "\n",
    "    # calculating co-expression matrix for cell type of interest\n",
    "    obj <- subset(object, subset = cluster_celltype == i)\n",
    "    \n",
    "    # only focus on expressed protein coding genes in the cell type of interes\n",
    "    counts <- LayerData(obj,assay = \"PC\",layer = \"counts\")# code in seurat 5.0+\n",
    "    genes.percent.expression <- rowMeans(counts>0 )*100   \n",
    "    genes_selected <- names(genes.percent.expression[genes.percent.expression>25])\n",
    "    message(paste(length(genes_selected),\" genes expressed in >25% cells.\"))\n",
    "\n",
    "    motif.res <- enriched.motif[enriched.motif$celltype == i,]\n",
    "    motif.res <- motif.res[motif.res$motif.name %in% genes_selected,]\n",
    "    message(length(motif.res$motif))\n",
    "\n",
    "    expr.res <-rbind(expr.res,motif.res)\n",
    "\n",
    "}\n",
    "\n",
    "table(expr.res$celltype,expr.res$deg_dir)\n",
    "## saving results\n",
    "write.csv(expr.res, file = \"./Results/LINK/PFC_enriched_JASPAR_motif_expr.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## running analysis on EC region object\n",
    "# load data\n",
    "object <- ec\n",
    "object\n",
    "\n",
    "## top upregulated genes\n",
    "degs <- read.csv(\"./Results/DEG/Overlap_mast_mixed_EC.csv\",row.names = 1)\n",
    "message(\"How many DEGs in each cell type in EC.\")\n",
    "dim(degs)\n",
    "table(degs$celltype,degs$dir)\n",
    "#degs\n",
    "\n",
    "## read in the link data\n",
    "links<- read.csv(\"./Results/LINK/EC_linkpeaks_all_annotated.csv\")\n",
    "links$new_comb <- paste(links$gene,links$deg_dir,links$celltype,sep = \"_\")\n",
    "\n",
    "#\n",
    "message(\"How many linked peaks in each cell type in EC.\")\n",
    "dim(links)\n",
    "table(links$celltype,links$deg_dir)\n",
    "\n",
    "#\n",
    "message(\"How many linked DEGs in each cell type in EC.\")\n",
    "degs <- degs[degs$comb %in% unique(links$new_comb),]\n",
    "table(degs$celltype,degs$dir)\n",
    "length(degs$comb)\n",
    "#degs\n",
    "links <- links[links$new_comb %in% degs$comb,]\n",
    "\n",
    "ct <- c(\"Astrocyte\",\"Excitatory\",\"Inhibitory\",\"Microglia\",\"Oligodendrocyte\",\"OPC\")\n",
    "direction <- c(\"pos\",\"neg\")\n",
    "Idents(object) <- \"cluster_celltype\"\n",
    "final <- data.frame()\n",
    "\n",
    " ## running\n",
    "for (i in ct){\n",
    "    for(j in direction){\n",
    "        celltype <- i\n",
    "        print(celltype)\n",
    "        deg_direction = j\n",
    "        print(dir)\n",
    "        \n",
    "        res1 <- get_motif(object=object, celltype=celltype, deg_direction = deg_direction)\n",
    "        print(length(res1$motif.name))\n",
    "        final <- rbind(final,res1)\n",
    "    }\n",
    "}\n",
    "\n",
    "#### generate heatmap for motif enriched results_EC\n",
    "enriched.motif <- final\n",
    "table(enriched.motif$celltype,enriched.motif$deg_dir)\n",
    "\n",
    "enriched.motif$motif.name <- str_split_fixed(enriched.motif$motif.name,\"\\\\.\",n = 3)[,3]\n",
    "enriched.motif$motif.name <- str_split_fixed(enriched.motif$motif.name,\"::\",n = 1)[,1]\n",
    "\n",
    "head(enriched.motif)\n",
    "\n",
    "## only keeping those TFs expressed in the specific cell type. \n",
    "expr.res <- data.frame()\n",
    "ct = c(\"Astrocyte\",\"Excitatory\",\"Inhibitory\",\"Microglia\",\"Oligodendrocyte\",\"OPC\")\n",
    "\n",
    "for (i in ct){\n",
    "    print(i)\n",
    "\n",
    "    # calculating co-expression matrix for cell type of interest\n",
    "    obj <- subset(object, subset = cluster_celltype == i)\n",
    "    \n",
    "    # only focus on expressed protein coding genes in the cell type of interes\n",
    "    counts <- LayerData(obj,assay = \"PC\",layer = \"counts\")# code in seurat 5.0+\n",
    "    genes.percent.expression <- rowMeans(counts>0 )*100   \n",
    "    genes_selected <- names(genes.percent.expression[genes.percent.expression>25])\n",
    "    message(paste(length(genes_selected),\" genes expressed in >25% cells.\"))\n",
    "\n",
    "    motif.res <- enriched.motif[enriched.motif$celltype == i,]\n",
    "    motif.res <- motif.res[motif.res$motif.name %in% genes_selected,]\n",
    "    message(length(motif.res$motif))\n",
    "\n",
    "    expr.res <-rbind(expr.res,motif.res)\n",
    "\n",
    "}\n",
    "\n",
    "table(expr.res$celltype,expr.res$deg_dir)\n",
    "## saving results\n",
    "write.csv(expr.res, file = \"./Results/LINK/EC_enriched_JASPAR_motif_expr.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## running analysis on HIP region object\n",
    "# load data\n",
    "object <- hip\n",
    "object\n",
    "\n",
    "## top upregulated genes\n",
    "degs <- read.csv(\"./Results/DEG/Overlap_mast_mixed_HIP.csv\",row.names = 1)\n",
    "message(\"How many DEGs in each cell type in HIP.\")\n",
    "dim(degs)\n",
    "table(degs$celltype,degs$dir)\n",
    "#degs\n",
    "\n",
    "## read in the link data\n",
    "links<- read.csv(\"./Results/LINK/HIP_linkpeaks_all_annotated.csv\")\n",
    "links$new_comb <- paste(links$gene,links$deg_dir,links$celltype,sep = \"_\")\n",
    "\n",
    "#\n",
    "message(\"How many linked peaks in each cell type in HIP.\")\n",
    "dim(links)\n",
    "table(links$celltype,links$deg_dir)\n",
    "\n",
    "#\n",
    "message(\"How many linked DEGs in each cell type in HIP.\")\n",
    "degs <- degs[degs$comb %in% unique(links$new_comb),]\n",
    "table(degs$celltype,degs$dir)\n",
    "length(degs$comb)\n",
    "#degs\n",
    "links <- links[links$new_comb %in% degs$comb,]\n",
    "\n",
    "## running\n",
    "ct <- c(\"Astrocyte\",\"Excitatory\",\"Inhibitory\",\"Microglia\",\"Oligodendrocyte\",\"OPC\")\n",
    "direction <- c(\"pos\",\"neg\")\n",
    "Idents(object) <- \"cluster_celltype\"\n",
    "final <- data.frame()\n",
    "\n",
    "#i = \"Excitatory\"\n",
    "#j = \"neg\" \n",
    "for (i in ct){\n",
    "    for(j in direction){\n",
    "        celltype <- i\n",
    "        print(celltype)\n",
    "        deg_direction = j\n",
    "        print(dir)\n",
    "        \n",
    "        res1 <- get_motif(object=object, celltype=celltype, deg_direction = deg_direction)\n",
    "        print(length(res1$motif.name))\n",
    "        final <- rbind(final,res1)\n",
    "    }\n",
    "}\n",
    "\n",
    "#### generate heatmap for motif enriched results_PFC\n",
    "enriched.motif <- final\n",
    "table(enriched.motif$celltype,enriched.motif$deg_dir)\n",
    "\n",
    "enriched.motif$motif.name <- str_split_fixed(enriched.motif$motif.name,\"\\\\.\",n = 3)[,3]\n",
    "enriched.motif$motif.name <- str_split_fixed(enriched.motif$motif.name,\"::\",n = 1)[,1]\n",
    "\n",
    "head(enriched.motif)\n",
    "\n",
    "## only keeping those TFs expressed in the specific cell type. \n",
    "expr.res <- data.frame()\n",
    "ct = c(\"Astrocyte\",\"Excitatory\",\"Inhibitory\",\"Microglia\",\"Oligodendrocyte\",\"OPC\")\n",
    "\n",
    "for (i in ct){\n",
    "    print(i)\n",
    "\n",
    "    # calculating co-expression matrix for cell type of interest\n",
    "    obj <- subset(object, subset = cluster_celltype == i)\n",
    "    \n",
    "    # only focus on expressed protein coding genes in the cell type of interes\n",
    "    counts <- LayerData(obj,assay = \"PC\",layer = \"counts\")# code in seurat 5.0+\n",
    "    genes.percent.expression <- rowMeans(counts>0 )*100   \n",
    "    genes_selected <- names(genes.percent.expression[genes.percent.expression>25])\n",
    "    message(paste(length(genes_selected),\" genes expressed in >25% cells.\"))\n",
    "\n",
    "    motif.res <- enriched.motif[enriched.motif$celltype == i,]\n",
    "    motif.res <- motif.res[motif.res$motif.name %in% genes_selected,]\n",
    "    message(length(motif.res$motif))\n",
    "\n",
    "    expr.res <-rbind(expr.res,motif.res)\n",
    "\n",
    "}\n",
    "\n",
    "table(expr.res$celltype,expr.res$deg_dir)\n",
    "\n",
    "## saving results\n",
    "write.csv(expr.res, file = \"./Results/LINK/HIP_enriched_JASPAR_motif_expr.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## visualization Figure 3e\n",
    "pfc_motif <- read.csv(\"/data2/aliu8/2023_AD_multiome/Analysis/Results/LINK/PFC_enriched_JASPAR_motif_expr.csv\",row.names = 1)\n",
    "pfc_motif$log2FC <- -log(pfc_motif$fold.enrichment,base = 2)\n",
    "ec_motif <- read.csv(\"/data2/aliu8/2023_AD_multiome/Analysis/Results/LINK/EC_enriched_JASPAR_motif_expr.csv\",row.names = 1)\n",
    "ec_motif$log2FC <- -log(ec_motif$fold.enrichment,base = 2)\n",
    "hip_motif <- read.csv(\"/data2/aliu8/2023_AD_multiome/Analysis/Results/LINK/HIP_enriched_JASPAR_motif_expr.csv\",row.names = 1)\n",
    "hip_motif$log2FC <- -log(hip_motif$fold.enrichment,base = 2)\n",
    "\n",
    "dim(pfc_motif)\n",
    "dim(ec_motif)\n",
    "dim(hip_motif)\n",
    "length(unique(c(pfc_motif$motif.name,ec_motif$motif.name,hip_motif$motif.name)))\n",
    "\n",
    "table(pfc_motif$celltype,pfc_motif$deg_dir)\n",
    "table(ec_motif$celltype,ec_motif$deg_dir)\n",
    "table(hip_motif$celltype,hip_motif$deg_dir)\n",
    "\n",
    "pfc_motif$region <- \"PFC\"\n",
    "ec_motif$region <- \"EC\"\n",
    "hip_motif$region <- \"HIP\"\n",
    "\n",
    "pfc_motif_pos <- pfc_motif[pfc_motif$deg_dir == \"pos\",]\n",
    "pfc_motif_neg <- pfc_motif[pfc_motif$deg_dir == \"neg\",]\n",
    "\n",
    "ec_motif_pos <- ec_motif[ec_motif$deg_dir == \"pos\",]\n",
    "ec_motif_neg <- ec_motif[ec_motif$deg_dir == \"neg\",]\n",
    "\n",
    "hip_motif_pos <- hip_motif[hip_motif$deg_dir == \"pos\",]\n",
    "hip_motif_neg <- hip_motif[hip_motif$deg_dir == \"neg\",]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# combine for plotting for upupregulated part\n",
    "motif_pos <- rbind(pfc_motif_pos,ec_motif_pos,hip_motif_pos)\n",
    "motif_pos$ct_rg <- paste(substr(motif_pos$celltype,start=1,stop=3),motif_pos$region, sep = \"_\")\n",
    "motif_pos$log10.adjp <- -log10(motif_pos$p.adjust)\n",
    "\n",
    "## order based on the following order\n",
    "cg_order <- c(\"Ast_PFC\",\"Ast_EC\",\"Ast_HIP\",\n",
    "                \"Exc_PFC\",\"Exc_EC\",\"Exc_HIP\",\n",
    "                \"Inh_PFC\",\"Inh_EC\",\"Inh_HIP\",\n",
    "                \"Mic_PFC\",\"Mic_EC\",\"Mic_HIP\",\n",
    "                \"Oli_PFC\",\"Oli_EC\",\"Oli_HIP\",\n",
    "                \"OPC_PFC\",\"OPC_EC\",\"OPC_HIP\")\n",
    "motif_pos <- motif_pos[order(match(motif_pos$ct_rg, cg_order)),]\n",
    "head(motif_pos)\n",
    "#motif_pos\n",
    "table(motif_pos$celltype, motif_pos$ct_rg)\n",
    "\n",
    "df <- motif_pos %>% group_by(ct_rg) %>% top_n(n = 10, wt = log10.adjp)\n",
    "#df\n",
    "length(unique(df$motif))\n",
    "top <- df$motif\n",
    "#top\n",
    "\n",
    "mat <- matrix(nrow=length(unique(top)), ncol=17)\n",
    "colnames(mat)<-unique(df$ct_rg)\n",
    "\n",
    "for (j in 1:length(unique(top))){\n",
    "  for (i in 1:17){\n",
    "    motifTemp<-unique(top)[j]\n",
    "    rgTemp<-colnames(mat)[i]\n",
    "\n",
    "    sub<-motif_pos[which(motif_pos$motif==motifTemp & motif_pos$ct_rg==rgTemp),]\n",
    "    if (nrow(sub)>0){\n",
    "      mat[j,i]<-sub$log10.adjp\n",
    "    }\n",
    "    else{\n",
    "      mat[j,i]<-0\n",
    "    }\n",
    "  }  \n",
    "}\n",
    "\n",
    "rownames(mat)<- unique(top)\n",
    "rownames(mat) <- str_split_fixed(rownames(mat),\"\\\\.\",n = 3)[,3]\n",
    "#mat\n",
    "\n",
    "sig_mat<-matrix(nrow=length(unique(top)), ncol=17)\n",
    "colnames(sig_mat)<-unique(df$ct_rg)\n",
    "\n",
    "for (j in 1:length(unique(top))){\n",
    "  for (i in 1:17){\n",
    "    motifTemp<-unique(top)[j]\n",
    "    rgTemp<-colnames(sig_mat)[i]\n",
    "\n",
    "    sub<-motif_pos[which(motif_pos$motif==motifTemp & motif_pos$ct_rg==rgTemp),]\n",
    "    if (nrow(sub)>0){\n",
    "      sig_mat[j,i]<-sub$p.adjust}\n",
    "    else{\n",
    "      sig_mat[j,i]<-1}  }}\n",
    "\n",
    "#sig_mat\n",
    "\n",
    "## plotting\n",
    "ha<-HeatmapAnnotation(Region=colnames(mat)\n",
    "                       , col= list(Region=c(\"Ast_PFC\"=\"#c25757ff\",\"Ast_EC\"=\"#825ca6ff\",\"Ast_HIP\"=\"#3f78c1ff\",\n",
    "                                            \"Exc_PFC\"=\"#c25757ff\",\"Exc_EC\"=\"#825ca6ff\",\"Exc_HIP\"=\"#3f78c1ff\",\n",
    "                                            \"Inh_PFC\"=\"#c25757ff\",\"Inh_EC\"=\"#825ca6ff\",\"Inh_HIP\"=\"#3f78c1ff\",\n",
    "                                            \"Mic_PFC\"=\"#c25757ff\",\"Mic_EC\"=\"#825ca6ff\",\"Mic_HIP\"=\"#3f78c1ff\",\n",
    "                                            \"Oli_PFC\"=\"#c25757ff\",\"Oli_EC\"=\"#825ca6ff\",\"Oli_HIP\"=\"#3f78c1ff\",\n",
    "                                            \"OPC_PFC\"=\"#c25757ff\",\"OPC_EC\"=\"#825ca6ff\",\"OPC_HIP\"=\"#3f78c1ff\")), show_legend=F,annotation_label=\"Brain region\")\n",
    "ha2<-HeatmapAnnotation(Celltype=colnames(mat)\n",
    "                       , col= list(Celltype=c(\"Ast_PFC\"=\"#F06719\",\"Ast_EC\"=\"#F06719\",\"Ast_HIP\"=\"#F06719\",\n",
    "                                            \"Exc_PFC\"=\"#33A65C\",\"Exc_EC\"=\"#33A65C\",\"Exc_HIP\"=\"#33A65C\",\n",
    "                                            \"Inh_PFC\"=\"#23767C\",\"Inh_EC\"=\"#23767C\",\"Inh_HIP\"=\"#23767C\",\n",
    "                                            \"Mic_PFC\"=\"#E03426\",\"Mic_EC\"=\"#E03426\",\"Mic_HIP\"=\"#E03426\",\n",
    "                                            \"Oli_PFC\"=\"#1BA3C6\",\"Oli_EC\"=\"#1BA3C6\",\"Oli_HIP\"=\"#1BA3C6\",\n",
    "                                            \"OPC_PFC\"=\"#A26DC2\",\"OPC_HIP\"=\"#A26DC2\",\"OPC_EC\"=\"#A26DC2\")), show_legend=F,annotation_label=\"Cell type\")\n",
    "ha <- c(ha2,ha)\n",
    "\n",
    "ht1 <- Heatmap(mat,\n",
    "    cluster_rows = F,\n",
    "    cluster_columns = F,\n",
    "    col = colorRamp2(c(0,50,100),c(\"grey99\",\"red\",\"red4\")),\n",
    "    row_names_side = \"left\",#row_names_gp = gpar(fontface=\"italic\"),\n",
    "    top_annotation=ha,show_column_names=F,show_row_dend = F,\n",
    "    cell_fun = function(j, i, x, y, w, h, fill) {\n",
    "  if(sig_mat[i, j] <0.05) {\n",
    "    grid.text(\"*\", x, y, gp=gpar(fontsize=15, col=\"black\"), vjust=\"center\")\n",
    "  } })\n",
    "\n",
    "\n",
    "\n",
    "## writing figures\n",
    "pdf(\"/data2/aliu8/2023_AD_multiome/Analysis/Figures/LINK/motif_enrichment_pos.pdf\",height = 8,width = 5.5)\n",
    "ht1\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# combine for plotting for the down regulated\n",
    "motif_neg <- rbind(pfc_motif_neg,ec_motif_neg,hip_motif_neg)\n",
    "motif_neg$ct_rg <- paste(substr(motif_neg$celltype,start=1,stop=3),motif_neg$region, sep = \"_\")\n",
    "motif_neg$log10.adjp <- -log10(motif_neg$p.adjust)\n",
    "\n",
    "table(motif_neg$celltype, motif_neg$region)\n",
    "## order based on the following order\n",
    "cg_order <- c(\"Ast_PFC\",\"Ast_EC\",\"Ast_HIP\",\n",
    "                \"Exc_PFC\",\"Exc_EC\",\"Exc_HIP\",\n",
    "                \"Inh_PFC\",\"Inh_EC\",\"Inh_HIP\",\n",
    "                \"Mic_PFC\",\"Mic_EC\",\"Mic_HIP\",\n",
    "                \"Oli_PFC\",\"Oli_EC\",\"Oli_HIP\",\n",
    "                \"OPC_PFC\",\"OPC_EC\",\"OPC_HIP\")\n",
    "\n",
    "motif_neg <- motif_neg[order(match(motif_neg$ct_rg, cg_order)),]\n",
    "head(motif_neg)\n",
    "#motif_pos\n",
    "table(motif_neg$celltype, motif_neg$region)\n",
    "\n",
    "df <- motif_neg %>% group_by(ct_rg) %>% top_n(n = 10, wt = log10.adjp)\n",
    "length(unique(df$motif))\n",
    "top <- df$motif\n",
    "\n",
    "#motif_neg\n",
    "\n",
    "mat <- matrix(nrow=length(unique(top)), ncol=18)\n",
    "colnames(mat)<-unique(df$ct_rg)\n",
    "\n",
    "for (j in 1:length(unique(top))){\n",
    "  for (i in 1:18){\n",
    "    motifTemp<-unique(top)[j]\n",
    "    rgTemp<-colnames(mat)[i]\n",
    "\n",
    "    sub<-motif_neg[which(motif_neg$motif==motifTemp & motif_neg$ct_rg==rgTemp),]\n",
    "    if (nrow(sub)>0){\n",
    "      mat[j,i]<-sub$log10.adjp\n",
    "    }\n",
    "    else{\n",
    "      mat[j,i]<-0\n",
    "    }\n",
    "  }  \n",
    "}\n",
    "\n",
    "rownames(mat)<- unique(top)\n",
    "rownames(mat) <- str_split_fixed(rownames(mat),\"\\\\.\",n = 3)[,3]\n",
    "#mat\n",
    "\n",
    "\n",
    "sig_mat<-matrix(nrow=length(unique(top)), ncol=18)\n",
    "colnames(sig_mat)<-unique(df$ct_rg)\n",
    "\n",
    "for (j in 1:length(unique(top))){\n",
    "  for (i in 1:18){\n",
    "    motifTemp<-unique(top)[j]\n",
    "    rgTemp<-colnames(sig_mat)[i]\n",
    "\n",
    "    sub<-motif_neg[which(motif_neg$motif==motifTemp & motif_neg$ct_rg==rgTemp),]\n",
    "    if (nrow(sub)>0){\n",
    "      sig_mat[j,i]<-sub$p.adjust}\n",
    "    else{\n",
    "      sig_mat[j,i]<-1}  }}\n",
    "\n",
    "\n",
    "#sig_mat\n",
    " \n",
    "## plotting\n",
    "ha<-HeatmapAnnotation(Region=colnames(mat)\n",
    "                       , col= list(Region=c(\"Ast_PFC\"=\"#c25757ff\",\"Ast_EC\"=\"#825ca6ff\",\"Ast_HIP\"=\"#3f78c1ff\",\n",
    "                                            \"Exc_PFC\"=\"#c25757ff\",\"Exc_EC\"=\"#825ca6ff\",\"Exc_HIP\"=\"#3f78c1ff\",\n",
    "                                            \"Inh_PFC\"=\"#c25757ff\",\"Inh_EC\"=\"#825ca6ff\",\"Inh_HIP\"=\"#3f78c1ff\",\n",
    "                                            \"Mic_PFC\"=\"#c25757ff\",\"Mic_EC\"=\"#825ca6ff\",\"Mic_HIP\"=\"#3f78c1ff\",\n",
    "                                            \"Oli_PFC\"=\"#c25757ff\",\"Oli_EC\"=\"#825ca6ff\",\"Oli_HIP\"=\"#3f78c1ff\",\n",
    "                                            \"OPC_PFC\"=\"#c25757ff\",\"OPC_EC\"=\"#825ca6ff\",\"OPC_HIP\"=\"#3f78c1ff\")), show_legend=F,annotation_label=\"Brain region\")\n",
    "ha2<-HeatmapAnnotation(Celltype=colnames(mat)\n",
    "                       , col= list(Celltype=c(\"Ast_PFC\"=\"#F06719\",\"Ast_EC\"=\"#F06719\",\"Ast_HIP\"=\"#F06719\",\n",
    "                                            \"Exc_PFC\"=\"#33A65C\",\"Exc_EC\"=\"#33A65C\",\"Exc_HIP\"=\"#33A65C\",\n",
    "                                            \"Inh_PFC\"=\"#23767C\",\"Inh_EC\"=\"#23767C\",\"Inh_HIP\"=\"#23767C\",\n",
    "                                            \"Mic_PFC\"=\"#E03426\",\"Mic_EC\"=\"#E03426\",\"Mic_HIP\"=\"#E03426\",\n",
    "                                            \"Oli_PFC\"=\"#1BA3C6\",\"Oli_EC\"=\"#1BA3C6\",\"Oli_HIP\"=\"#1BA3C6\",\n",
    "                                            \"OPC_PFC\"=\"#A26DC2\",\"OPC_HIP\"=\"#A26DC2\",\"OPC_EC\"=\"#A26DC2\")), show_legend=F,annotation_label=\"Cell type\")\n",
    "ha <- c(ha2,ha)\n",
    "\n",
    "ht2 <- Heatmap(mat,\n",
    "    cluster_rows = F,\n",
    "    cluster_columns = F,\n",
    "    col = colorRamp2(c(0,50,100),c(\"grey99\",\"deepskyblue\",\"dodgerblue4\")),\n",
    "    row_names_side = \"left\", #row_names_gp = gpar(fontface=\"italic\"),\n",
    "    top_annotation=ha,show_column_names=F,show_row_dend = F,\n",
    "      cell_fun = function(j, i, x, y, w, h, fill) {\n",
    "  if(sig_mat[i, j] <0.05) {\n",
    "    grid.text(\"*\", x, y, gp=gpar(fontsize=15, col=\"black\"), vjust=\"center\")\n",
    "  } }\n",
    "    )\n",
    "\n",
    "## saving\n",
    "pdf(\"./Figures/LINK/motif_enrichment_neg.pdf\",height = 8,width = 6)\n",
    "ht2\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "### Figure 3/4: footprinting analysis \n",
    "TF_oi <- read.csv(\"./Results/LINK/TF_oi.csv\")\n",
    "\n",
    "## working on EC first\n",
    "object <- ec\n",
    "## running loop\n",
    "for (i in 1:length(TF_oi$motif)){\n",
    "  print(i)\n",
    "  tf.name <- TF_oi$motif[i]\n",
    "  \n",
    "  outpath <- paste(\"./Figures/LINK/EC/EC_footprinting_\",TF_oi$motif.name[i],\".pdf\",sep = \"\")\n",
    "  \n",
    "  # plotting\n",
    "  ## footprinting analysis visualization\n",
    "  object$cell_diag <- paste(object$cluster_celltype, object$diagnosis, sep = \"_\")\n",
    "  Idents(object) <- object$cell_diag\n",
    "  levels(object) <- sort(levels(object))\n",
    "  \n",
    "  pdf(outpath,width = 6,height = 5)\n",
    "  \n",
    "  print(PlotFootprint(object, features = tf.name,label = F,idents = c(\"Astrocyte_EOAD\",\"Astrocyte_NCI\",\"Excitatory_EOAD\",\"Excitatory_NCI\",\"Inhibitory_NCI\",\"Inhibitory_EOAD\",\"Microglia_EOAD\",\"Microglia_NCI\",\"Oligodendrocyte_EOAD\",\"Oligodendrocyte_NCI\",\"OPC_EOAD\",\"OPC_NCI\")))\n",
    "  print(PlotFootprint(object, features = tf.name,label = F,idents = c(\"Astrocyte_EOAD\",\"Astrocyte_NCI\")))\n",
    "  print(PlotFootprint(object, features = tf.name,label = F,idents = c(\"Excitatory_EOAD\",\"Excitatory_NCI\")))\n",
    "  print(PlotFootprint(object, features = tf.name,label = F,idents = c(\"Inhibitory_NCI\",\"Inhibitory_EOAD\")))\n",
    "  print(PlotFootprint(object, features = tf.name,label = F,idents = c(\"Microglia_NCI\",\"Microglia_EOAD\")))\n",
    "  print(PlotFootprint(object, features = tf.name,label = F,idents = c(\"Oligodendrocyte_EOAD\",\"Oligodendrocyte_NCI\")))\n",
    "  print(PlotFootprint(object, features = tf.name,label = F,idents = c(\"OPC_EOAD\",\"OPC_NCI\")))\n",
    "  \n",
    "  dev.off()\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "### Figure 3/4: footprinting analysis \n",
    "## HIP\n",
    "object <- hip\n",
    "\n",
    "## running loop\n",
    "for (i in 1:length(TF_oi$motif)){\n",
    "  print(i)\n",
    "  tf.name <- TF_oi$motif[i]\n",
    "  \n",
    "  outpath <- paste(\"./Figures/LINK/HIP/HIP_footprinting_\",TF_oi$motif.name[i],\".pdf\",sep = \"\")\n",
    "  \n",
    "  # plotting\n",
    "  ## footprinting analysis visualization\n",
    "  object$cell_diag <- paste(object$cluster_celltype, object$diagnosis, sep = \"_\")\n",
    "  Idents(object) <- object$cell_diag\n",
    "  levels(object) <- sort(levels(object))\n",
    "  \n",
    "  pdf(outpath,width = 6,height = 5)\n",
    "  \n",
    "  print(PlotFootprint(object, features = tf.name,label = F,idents = c(\"Astrocyte_EOAD\",\"Astrocyte_NCI\",\"Excitatory_EOAD\",\"Excitatory_NCI\",\"Inhibitory_NCI\",\"Inhibitory_EOAD\",\"Microglia_EOAD\",\"Microglia_NCI\",\"Oligodendrocyte_EOAD\",\"Oligodendrocyte_NCI\",\"OPC_EOAD\",\"OPC_NCI\")))\n",
    "  print(PlotFootprint(object, features = tf.name,label = F,idents = c(\"Astrocyte_EOAD\",\"Astrocyte_NCI\")))\n",
    "  print(PlotFootprint(object, features = tf.name,label = F,idents = c(\"Excitatory_EOAD\",\"Excitatory_NCI\")))\n",
    "  print(PlotFootprint(object, features = tf.name,label = F,idents = c(\"Inhibitory_NCI\",\"Inhibitory_EOAD\")))\n",
    "  print(PlotFootprint(object, features = tf.name,label = F,idents = c(\"Microglia_NCI\",\"Microglia_EOAD\")))\n",
    "  print(PlotFootprint(object, features = tf.name,label = F,idents = c(\"Oligodendrocyte_EOAD\",\"Oligodendrocyte_NCI\")))\n",
    "  print(PlotFootprint(object, features = tf.name,label = F,idents = c(\"OPC_EOAD\",\"OPC_NCI\")))\n",
    "  \n",
    "  dev.off()\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "### Figure 3/4: footprinting analysis \n",
    "## PFC\n",
    "object <- pfc\n",
    "\n",
    "## running loop\n",
    "for (i in 1:length(TF_oi$motif)){\n",
    "  print(i)\n",
    "  tf.name <- TF_oi$motif[i]\n",
    "  \n",
    "  outpath <- paste(\"/data2/aliu8/2023_AD_multiome/Analysis/Figures/LINK/PFC/PFC_footprinting_\",TF_oi$motif.name[i],\".pdf\",sep = \"\")\n",
    "  \n",
    "  # plotting\n",
    "  ## footprinting analysis visualization\n",
    "  object$cell_diag <- paste(object$cluster_celltype, object$diagnosis, sep = \"_\")\n",
    "  Idents(object) <- object$cell_diag\n",
    "  levels(object) <- sort(levels(object))\n",
    "  \n",
    "  pdf(outpath,width = 6,height = 5)\n",
    "  \n",
    "  print(PlotFootprint(object, features = tf.name,label = F,idents = c(\"Astrocyte_EOAD\",\"Astrocyte_NCI\",\"Excitatory_EOAD\",\"Excitatory_NCI\",\"Inhibitory_NCI\",\"Inhibitory_EOAD\",\"Microglia_EOAD\",\"Microglia_NCI\",\"Oligodendrocyte_EOAD\",\"Oligodendrocyte_NCI\",\"OPC_EOAD\",\"OPC_NCI\")))\n",
    "  print(PlotFootprint(object, features = tf.name,label = F,idents = c(\"Astrocyte_EOAD\",\"Astrocyte_NCI\")))\n",
    "  print(PlotFootprint(object, features = tf.name,label = F,idents = c(\"Excitatory_EOAD\",\"Excitatory_NCI\")))\n",
    "  print(PlotFootprint(object, features = tf.name,label = F,idents = c(\"Inhibitory_NCI\",\"Inhibitory_EOAD\")))\n",
    "  print(PlotFootprint(object, features = tf.name,label = F,idents = c(\"Microglia_NCI\",\"Microglia_EOAD\")))\n",
    "  print(PlotFootprint(object, features = tf.name,label = F,idents = c(\"Oligodendrocyte_EOAD\",\"Oligodendrocyte_NCI\")))\n",
    "  print(PlotFootprint(object, features = tf.name,label = F,idents = c(\"OPC_EOAD\",\"OPC_NCI\")))\n",
    "  \n",
    "  dev.off()\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "####################################\n",
    "## Working on the SCENIC+ results ##\n",
    "####################################\n",
    "\n",
    "## first, filter the regulons\n",
    "c1 <- read.csv(\"/data2/ccitu/andi_multiome_filtered/eregulon_files_22_03_2024/tf_cistrome_corr_pfc.csv\",row.names = 1)\n",
    "c2 <- read.csv(\"/data2/ccitu/andi_multiome_filtered/eregulon_files_22_03_2024/tf_cistrome_corr_ec.csv\",row.names = 1)\n",
    "c3 <- read.csv(\"/data2/ccitu/andi_multiome_filtered/eregulon_files_22_03_2024/tf_cistrome_corr_hip.csv\",row.names = 1)\n",
    "\n",
    "### select the top 5% positive correlation and bottom 5% negetive correlation\n",
    "c1 <- c1[c1$Rho <= -0.66 | c1$Rho >= 0.89,]\n",
    "c2 <- c2[c2$Rho <= -0.71 | c2$Rho >= 0.86,]\n",
    "c3 <- c3[c3$Rho <= -0.78 | c3$Rho >= 0.93,]\n",
    "\n",
    "dim(c1)\n",
    "dim(c2)\n",
    "dim(c3)\n",
    "\n",
    "intersect(intersect(c1$TF,c2$TF),c3$TF)\n",
    "\n",
    "write.csv(c1, file = \"./Results/SCENIC/PFC_regulon.csv\",row.names = F)\n",
    "write.csv(c2, file = \"./Results/SCENIC/EC_regulon.csv\",row.names = F)\n",
    "write.csv(c3, file = \"./Results/SCENIC/HIP_regulon.csv\",row.names = F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "### Supplementar Figure Overlap TFs predicted by SCENIC+ across three brain regions\n",
    "## drawing venn diagram\n",
    "library(VennDiagram)\n",
    "library(RColorBrewer)\n",
    "myCol <- c(\"#C25757\", \"#825ca6\", \"#3F78C1\")#brewer.pal(3, \"Pastel2\")\n",
    "\n",
    "list1 <- unique(c1$TF)\n",
    "list2 <- unique(c2$TF)\n",
    "list3 <- unique(c3$TF)\n",
    "\n",
    "venn.diagram(\n",
    "  x = list(list1, list2, list3),\n",
    "  category.names = c(\"PFC\" , \"EC\" , \"HIP\"),\n",
    "  filename = './Figures/GRN/Venn_SCENIC+_ALL.png',\n",
    "  output=F,\n",
    "  # Output features\n",
    "  imagetype=\"png\" ,\n",
    "  height = 550, \n",
    "  width = 550, \n",
    "  resolution = 300,\n",
    "  compression = \"lzw\",\n",
    "  \n",
    "  # Circles\n",
    "  lwd = 2,\n",
    "  lty = 'blank',\n",
    "  fill = myCol,\n",
    "  \n",
    "  # Numbers\n",
    "  cex = .6,\n",
    "  fontface = \"bold\",\n",
    "  fontfamily = \"sans\",\n",
    "  \n",
    "  # Set names\n",
    "  cat.cex = 0.6,\n",
    "  cat.fontface = \"bold\",\n",
    "  cat.default.pos = \"outer\",\n",
    "#  cat.pos = c(-27, 27, 135),\n",
    "#  cat.dist = c(0.055, 0.055, 0.085),\n",
    "  cat.fontfamily = \"sans\"\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#################################################\n",
    "## creating jaccard score matrix on gene level ##\n",
    "#################################################\n",
    "library(ComplexHeatmap)\n",
    "library(circlize)\n",
    "library(viridis)\n",
    "## all edges from scenic+\n",
    "## specific region\n",
    "scenic <- read.csv(\"./Results/SCENIC_plus/eregulons_hip.csv\")\n",
    "## cistrom\n",
    "cistrome <- read.csv(\"./Results/SCENIC_plus/tf_cistrome_corr_hip.csv\")\n",
    "cistrome <- cistrome[cistrome$Rho >= 0.93 | cistrome$Rho <= -0.78,]# hip\n",
    "#cistrome <- cistrome[cistrome$Rho >= 0.89 | cistrome$Rho <= -0.66,] # pfc\n",
    "#cistrome <- cistrome[cistrome$Rho >= 0.86 | cistrome$Rho <= -0.71,] # ec\n",
    "\n",
    "j_mat <- matrix(nrow=length(unique(cistrome$Cistrome)),ncol = length(unique(cistrome$Cistrome)))\n",
    "rownames(j_mat) <- unique(cistrome$Cistrome)\n",
    "colnames(j_mat) <- unique(cistrome$Cistrome)\n",
    "#j_mat\n",
    "\n",
    "for (i in unique(cistrome$Cistrome)){\n",
    "    for (j in unique(cistrome$Cistrome)){\n",
    "        gene_list_i <- unique(scenic[scenic$Region_signature_name == i,]$Gene)\n",
    "        gene_list_j <- unique(scenic[scenic$Region_signature_name == j,]$Gene)\n",
    "        \n",
    "        j_mat[i,j]<- length(intersect(gene_list_i,gene_list_j))/length(union(gene_list_i,gene_list_j))\n",
    "        \n",
    "    }\n",
    "\n",
    "}\n",
    "\n",
    "pdf(file = \"./Figures/GRN/HIP_SCENIC_Gene_Jaccard.pdf\", height = 6,width = 8)\n",
    "p1 <- Heatmap(\n",
    "    j_mat,cluster_rows = T,\n",
    "    cluster_columns = T, \n",
    "    show_column_names=F,\n",
    "    show_row_dend = F,\n",
    "    show_column_dend = F,\n",
    "    col = colorRamp2(c(0,0.25,0.5),viridis(3)),row_names_side = \"left\")\n",
    "p1\n",
    "dev.off()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
